{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Covid19 Semantic-based Search using Word Embedding\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Goal\n",
    "\n",
    "\n",
    "With over 57,000 scholarly articles on the coronavirus family, it is extremely difficult for the medical researchers to go through this tremendous amount of research papers, hence very difficult to get useful insights about the new Covid-19 pandemic. **The main goal** is to implement a **semantic-based search** rather a *keyword-based* search.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Approach\n",
    "\n",
    "\n",
    "Instead of comparing occurences and counts, we will use *gensim's word2vec* in order to generate word embedding using the abstract texts as our corpus. For each document, we calculate the centroid of its abstract and for each query word, we map it to a vector then calculate the word centroid similarity for the query and each document's abstract. The top ranked papers are then selected and output."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Dataset Loading and Preprocessing\n",
    "\n",
    "\n",
    "Some loading and pre-processing steps are introduced by the notebooks by ****Ivan Ega Pratama****, ****Maksim Ekin**** from Kaggle. \n",
    "\n",
    "\n",
    "\n",
    "**Citation: ** [COVID EDA: Initial Exploration Tool](https://www.kaggle.com/ivanegapratama/covid-eda-initial-exploration-tool) and [COVID-19 Literature Clustering](https://www.kaggle.com/maksimeren/covid-19-literature-clustering#Loading-the-Data)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Since the data is too large to work with, we will use the abstracts only. We will start by loading the metadata file and extract from it the paper_id, title, abstract and doi."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: langdetect in c:\\users\\maggi\\anaconda3\\envs\\nlp\\lib\\site-packages (1.0.8)\n",
      "Requirement already satisfied: six in c:\\users\\maggi\\anaconda3\\envs\\nlp\\lib\\site-packages (from langdetect) (1.14.0)\n"
     ]
    }
   ],
   "source": [
    "# Imports \n",
    "!pip install langdetect\n",
    "\n",
    "import spacy\n",
    "import string\n",
    "import warnings\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "\n",
    "from pprint import pprint\n",
    "from IPython.utils import io\n",
    "from tqdm.notebook import tqdm\n",
    "from gensim.models import Word2Vec\n",
    "from langdetect import DetectorFactory, detect\n",
    "from IPython.core.display import HTML, display\n",
    "from spacy.lang.en.stop_words import STOP_WORDS\n",
    "\n",
    "warnings.filterwarnings('ignore')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "_cell_guid": "79c7e3d0-c299-4dcb-8224-4455121ee9b0",
    "_uuid": "d629ff2d2480ee46fbb7e2d37f6b5fab8052498a"
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cord_uid</th>\n",
       "      <th>sha</th>\n",
       "      <th>source_x</th>\n",
       "      <th>title</th>\n",
       "      <th>doi</th>\n",
       "      <th>pmcid</th>\n",
       "      <th>pubmed_id</th>\n",
       "      <th>license</th>\n",
       "      <th>abstract</th>\n",
       "      <th>publish_time</th>\n",
       "      <th>authors</th>\n",
       "      <th>journal</th>\n",
       "      <th>Microsoft Academic Paper ID</th>\n",
       "      <th>WHO #Covidence</th>\n",
       "      <th>has_pdf_parse</th>\n",
       "      <th>has_pmc_xml_parse</th>\n",
       "      <th>full_text_file</th>\n",
       "      <th>url</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>xqhn0vbp</td>\n",
       "      <td>1e1286db212100993d03cc22374b624f7caee956</td>\n",
       "      <td>PMC</td>\n",
       "      <td>Airborne rhinovirus detection and effect of ul...</td>\n",
       "      <td>10.1186/1471-2458-3-5</td>\n",
       "      <td>PMC140314</td>\n",
       "      <td>12525263</td>\n",
       "      <td>no-cc</td>\n",
       "      <td>BACKGROUND: Rhinovirus, the most common cause ...</td>\n",
       "      <td>2003-01-13</td>\n",
       "      <td>Myatt, Theodore A; Johnston, Sebastian L; Rudn...</td>\n",
       "      <td>BMC Public Health</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>custom_license</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/pmc/articles/PMC1...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>gi6uaa83</td>\n",
       "      <td>8ae137c8da1607b3a8e4c946c07ca8bda67f88ac</td>\n",
       "      <td>PMC</td>\n",
       "      <td>Discovering human history from stomach bacteria</td>\n",
       "      <td>10.1186/gb-2003-4-5-213</td>\n",
       "      <td>PMC156578</td>\n",
       "      <td>12734001</td>\n",
       "      <td>no-cc</td>\n",
       "      <td>Recent analyses of human pathogens have reveal...</td>\n",
       "      <td>2003-04-28</td>\n",
       "      <td>Disotell, Todd R</td>\n",
       "      <td>Genome Biol</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>custom_license</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/pmc/articles/PMC1...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>le0ogx1s</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PMC</td>\n",
       "      <td>A new recruit for the army of the men of death</td>\n",
       "      <td>10.1186/gb-2003-4-7-113</td>\n",
       "      <td>PMC193621</td>\n",
       "      <td>12844350</td>\n",
       "      <td>no-cc</td>\n",
       "      <td>The army of the men of death, in John Bunyan's...</td>\n",
       "      <td>2003-06-27</td>\n",
       "      <td>Petsko, Gregory A</td>\n",
       "      <td>Genome Biol</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>custom_license</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/pmc/articles/PMC1...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>fy4w7xz8</td>\n",
       "      <td>0104f6ceccf92ae8567a0102f89cbb976969a774</td>\n",
       "      <td>PMC</td>\n",
       "      <td>Association of HLA class I with severe acute r...</td>\n",
       "      <td>10.1186/1471-2350-4-9</td>\n",
       "      <td>PMC212558</td>\n",
       "      <td>12969506</td>\n",
       "      <td>no-cc</td>\n",
       "      <td>BACKGROUND: The human leukocyte antigen (HLA) ...</td>\n",
       "      <td>2003-09-12</td>\n",
       "      <td>Lin, Marie; Tseng, Hsiang-Kuang; Trejaut, Jean...</td>\n",
       "      <td>BMC Med Genet</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>custom_license</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/pmc/articles/PMC2...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0qaoam29</td>\n",
       "      <td>5b68a553a7cbbea13472721cd1ad617d42b40c26</td>\n",
       "      <td>PMC</td>\n",
       "      <td>A double epidemic model for the SARS propagation</td>\n",
       "      <td>10.1186/1471-2334-3-19</td>\n",
       "      <td>PMC222908</td>\n",
       "      <td>12964944</td>\n",
       "      <td>no-cc</td>\n",
       "      <td>BACKGROUND: An epidemic of a Severe Acute Resp...</td>\n",
       "      <td>2003-09-10</td>\n",
       "      <td>Ng, Tuen Wai; Turinici, Gabriel; Danchin, Antoine</td>\n",
       "      <td>BMC Infect Dis</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>custom_license</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/pmc/articles/PMC2...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   cord_uid                                       sha source_x  \\\n",
       "0  xqhn0vbp  1e1286db212100993d03cc22374b624f7caee956      PMC   \n",
       "1  gi6uaa83  8ae137c8da1607b3a8e4c946c07ca8bda67f88ac      PMC   \n",
       "2  le0ogx1s                                       NaN      PMC   \n",
       "3  fy4w7xz8  0104f6ceccf92ae8567a0102f89cbb976969a774      PMC   \n",
       "4  0qaoam29  5b68a553a7cbbea13472721cd1ad617d42b40c26      PMC   \n",
       "\n",
       "                                               title                      doi  \\\n",
       "0  Airborne rhinovirus detection and effect of ul...    10.1186/1471-2458-3-5   \n",
       "1    Discovering human history from stomach bacteria  10.1186/gb-2003-4-5-213   \n",
       "2     A new recruit for the army of the men of death  10.1186/gb-2003-4-7-113   \n",
       "3  Association of HLA class I with severe acute r...    10.1186/1471-2350-4-9   \n",
       "4   A double epidemic model for the SARS propagation   10.1186/1471-2334-3-19   \n",
       "\n",
       "       pmcid pubmed_id license  \\\n",
       "0  PMC140314  12525263   no-cc   \n",
       "1  PMC156578  12734001   no-cc   \n",
       "2  PMC193621  12844350   no-cc   \n",
       "3  PMC212558  12969506   no-cc   \n",
       "4  PMC222908  12964944   no-cc   \n",
       "\n",
       "                                            abstract publish_time  \\\n",
       "0  BACKGROUND: Rhinovirus, the most common cause ...   2003-01-13   \n",
       "1  Recent analyses of human pathogens have reveal...   2003-04-28   \n",
       "2  The army of the men of death, in John Bunyan's...   2003-06-27   \n",
       "3  BACKGROUND: The human leukocyte antigen (HLA) ...   2003-09-12   \n",
       "4  BACKGROUND: An epidemic of a Severe Acute Resp...   2003-09-10   \n",
       "\n",
       "                                             authors            journal  \\\n",
       "0  Myatt, Theodore A; Johnston, Sebastian L; Rudn...  BMC Public Health   \n",
       "1                                   Disotell, Todd R        Genome Biol   \n",
       "2                                  Petsko, Gregory A        Genome Biol   \n",
       "3  Lin, Marie; Tseng, Hsiang-Kuang; Trejaut, Jean...      BMC Med Genet   \n",
       "4  Ng, Tuen Wai; Turinici, Gabriel; Danchin, Antoine     BMC Infect Dis   \n",
       "\n",
       "  Microsoft Academic Paper ID WHO #Covidence  has_pdf_parse  \\\n",
       "0                         NaN            NaN           True   \n",
       "1                         NaN            NaN           True   \n",
       "2                         NaN            NaN          False   \n",
       "3                         NaN            NaN           True   \n",
       "4                         NaN            NaN           True   \n",
       "\n",
       "   has_pmc_xml_parse  full_text_file  \\\n",
       "0               True  custom_license   \n",
       "1               True  custom_license   \n",
       "2               True  custom_license   \n",
       "3               True  custom_license   \n",
       "4               True  custom_license   \n",
       "\n",
       "                                                 url  \n",
       "0  https://www.ncbi.nlm.nih.gov/pmc/articles/PMC1...  \n",
       "1  https://www.ncbi.nlm.nih.gov/pmc/articles/PMC1...  \n",
       "2  https://www.ncbi.nlm.nih.gov/pmc/articles/PMC1...  \n",
       "3  https://www.ncbi.nlm.nih.gov/pmc/articles/PMC2...  \n",
       "4  https://www.ncbi.nlm.nih.gov/pmc/articles/PMC2...  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "root_path = 'kaggle/input/CORD-19-research-challenge'\n",
    "metadata_path = f'{root_path}/metadata.csv'\n",
    "meta_df = pd.read_csv(metadata_path, dtype={\n",
    "    'pubmed_id': str,\n",
    "    'Microsoft Academic Paper ID': str, \n",
    "    'doi': str,\n",
    "    'abstract': str\n",
    "})\n",
    "meta_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>paper_id</th>\n",
       "      <th>title</th>\n",
       "      <th>abstract</th>\n",
       "      <th>doi</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>b2897e1277f56641193a6db73825f707eed3e4c9</td>\n",
       "      <td>Sequence requirements for RNA strand transfer ...</td>\n",
       "      <td>Nidovirus subgenomic mRNAs contain a leader se...</td>\n",
       "      <td>10.1093/emboj/20.24.7220</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>e3d0d482ebd9a8ba81c254cc433f314142e72174</td>\n",
       "      <td>Crystal structure of murine sCEACAM1a[1,4]: a ...</td>\n",
       "      <td>CEACAM1 is a member of the carcinoembryonic an...</td>\n",
       "      <td>10.1093/emboj/21.9.2076</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>00b1d99e70f779eb4ede50059db469c65e8c1469</td>\n",
       "      <td>Synthesis of a novel hepatitis C virus protein...</td>\n",
       "      <td>Hepatitis C virus (HCV) is an important human ...</td>\n",
       "      <td>10.1093/emboj/20.14.3840</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>cf584e00f637cbd8f1bb35f3f09f5ed07b71aeb0</td>\n",
       "      <td>Structure of coronavirus main proteinase revea...</td>\n",
       "      <td>The key enzyme in coronavirus polyprotein proc...</td>\n",
       "      <td>10.1093/emboj/cdf327</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>dde02f11923815e6a16a31dd6298c46b109c5dfa</td>\n",
       "      <td>Discontinuous and non-discontinuous subgenomic...</td>\n",
       "      <td>Arteri-, corona-, toro- and roniviruses are ev...</td>\n",
       "      <td>10.1093/emboj/cdf635</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                   paper_id  \\\n",
       "0  b2897e1277f56641193a6db73825f707eed3e4c9   \n",
       "1  e3d0d482ebd9a8ba81c254cc433f314142e72174   \n",
       "2  00b1d99e70f779eb4ede50059db469c65e8c1469   \n",
       "3  cf584e00f637cbd8f1bb35f3f09f5ed07b71aeb0   \n",
       "4  dde02f11923815e6a16a31dd6298c46b109c5dfa   \n",
       "\n",
       "                                               title  \\\n",
       "0  Sequence requirements for RNA strand transfer ...   \n",
       "1  Crystal structure of murine sCEACAM1a[1,4]: a ...   \n",
       "2  Synthesis of a novel hepatitis C virus protein...   \n",
       "3  Structure of coronavirus main proteinase revea...   \n",
       "4  Discontinuous and non-discontinuous subgenomic...   \n",
       "\n",
       "                                            abstract                       doi  \n",
       "0  Nidovirus subgenomic mRNAs contain a leader se...  10.1093/emboj/20.24.7220  \n",
       "1  CEACAM1 is a member of the carcinoembryonic an...   10.1093/emboj/21.9.2076  \n",
       "2  Hepatitis C virus (HCV) is an important human ...  10.1093/emboj/20.14.3840  \n",
       "3  The key enzyme in coronavirus polyprotein proc...      10.1093/emboj/cdf327  \n",
       "4  Arteri-, corona-, toro- and roniviruses are ev...      10.1093/emboj/cdf635  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_covid = pd.DataFrame(columns=['paper_id', 'title','abstract', 'doi'])\n",
    "df_covid['paper_id'] = meta_df.sha\n",
    "df_covid['title'] = meta_df.title\n",
    "df_covid['abstract'] = meta_df.abstract\n",
    "df_covid['doi'] = meta_df.doi\n",
    "\n",
    "df_covid.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Duplicates and Null values.\n",
    "\n",
    "We will look into the data and check if we have any null values."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "RangeIndex: 59887 entries, 0 to 59886\n",
      "Data columns (total 4 columns):\n",
      "paper_id    45763 non-null object\n",
      "title       59724 non-null object\n",
      "abstract    48757 non-null object\n",
      "doi         55801 non-null object\n",
      "dtypes: object(4)\n",
      "memory usage: 1.8+ MB\n"
     ]
    }
   ],
   "source": [
    "df_covid.info()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "Int64Index: 38584 entries, 0 to 59886\n",
      "Data columns (total 4 columns):\n",
      "paper_id    38584 non-null object\n",
      "title       38584 non-null object\n",
      "abstract    38584 non-null object\n",
      "doi         38584 non-null object\n",
      "dtypes: object(4)\n",
      "memory usage: 1.5+ MB\n"
     ]
    }
   ],
   "source": [
    "df_covid.drop_duplicates(['abstract'], inplace=True)\n",
    "df_covid.dropna(inplace=True)\n",
    "df_covid.info()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Dropping non-English articles.\n",
    "\n",
    "Now we dropped the null values, and removed the duplicates as well. Now we will check the number of non-english articles and see if we can drop them for the sake of simplicity."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "854fef8104aa4bfa9ba033f34545ac30",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "HBox(children=(FloatProgress(value=0.0, max=38584.0), HTML(value='')))"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# set seed\n",
    "DetectorFactory.seed = 0\n",
    "\n",
    "# hold label - language\n",
    "languages = []\n",
    "\n",
    "# go through each text\n",
    "for ii in tqdm(range(0,len(df_covid))):\n",
    "    # split by space into list, take the first x intex, join with space\n",
    "    text = df_covid.iloc[ii]['abstract'].split(\" \")\n",
    "    \n",
    "    lang = \"en\"\n",
    "    try:\n",
    "        if len(text) > 50:\n",
    "            lang = detect(\" \".join(text[:50]))\n",
    "        elif len(text) > 0:\n",
    "            lang = detect(\" \".join(text[:len(text)]))\n",
    "    # ught... beginning of the document was not in a good format\n",
    "    except Exception as e:\n",
    "        all_words = set(text)\n",
    "        try:\n",
    "            lang = detect(\" \".join(all_words))\n",
    "        \n",
    "        except Exception as e:        \n",
    "            lang = \"unknown\"\n",
    "            pass\n",
    "    \n",
    "    # get the language    \n",
    "    languages.append(lang)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's look at the numbers of articles for each language."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total: 38584\n",
      "\n",
      "{'af': 1,\n",
      " 'ca': 4,\n",
      " 'de': 57,\n",
      " 'en': 38071,\n",
      " 'es': 187,\n",
      " 'et': 1,\n",
      " 'fr': 205,\n",
      " 'it': 11,\n",
      " 'nl': 43,\n",
      " 'pt': 2,\n",
      " 'unknown': 2}\n"
     ]
    }
   ],
   "source": [
    "languages_dict = {}\n",
    "for lang in set(languages):\n",
    "    languages_dict[lang] = languages.count(lang)\n",
    "    \n",
    "print(\"Total: {}\\n\".format(len(languages)))\n",
    "pprint(languages_dict)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Since most of the articles are in English, we can safely drop non-English articles."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "Int64Index: 38071 entries, 0 to 59886\n",
      "Data columns (total 5 columns):\n",
      "paper_id    38071 non-null object\n",
      "title       38071 non-null object\n",
      "abstract    38071 non-null object\n",
      "doi         38071 non-null object\n",
      "language    38071 non-null object\n",
      "dtypes: object(5)\n",
      "memory usage: 1.7+ MB\n"
     ]
    }
   ],
   "source": [
    "df_covid['language'] = languages\n",
    "df_covid = df_covid[df_covid['language'] == 'en'] \n",
    "df_covid.info()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>paper_id</th>\n",
       "      <th>title</th>\n",
       "      <th>abstract</th>\n",
       "      <th>doi</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>b2897e1277f56641193a6db73825f707eed3e4c9</td>\n",
       "      <td>Sequence requirements for RNA strand transfer ...</td>\n",
       "      <td>Nidovirus subgenomic mRNAs contain a leader se...</td>\n",
       "      <td>10.1093/emboj/20.24.7220</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>e3d0d482ebd9a8ba81c254cc433f314142e72174</td>\n",
       "      <td>Crystal structure of murine sCEACAM1a[1,4]: a ...</td>\n",
       "      <td>CEACAM1 is a member of the carcinoembryonic an...</td>\n",
       "      <td>10.1093/emboj/21.9.2076</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>00b1d99e70f779eb4ede50059db469c65e8c1469</td>\n",
       "      <td>Synthesis of a novel hepatitis C virus protein...</td>\n",
       "      <td>Hepatitis C virus (HCV) is an important human ...</td>\n",
       "      <td>10.1093/emboj/20.14.3840</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>cf584e00f637cbd8f1bb35f3f09f5ed07b71aeb0</td>\n",
       "      <td>Structure of coronavirus main proteinase revea...</td>\n",
       "      <td>The key enzyme in coronavirus polyprotein proc...</td>\n",
       "      <td>10.1093/emboj/cdf327</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>dde02f11923815e6a16a31dd6298c46b109c5dfa</td>\n",
       "      <td>Discontinuous and non-discontinuous subgenomic...</td>\n",
       "      <td>Arteri-, corona-, toro- and roniviruses are ev...</td>\n",
       "      <td>10.1093/emboj/cdf635</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                   paper_id  \\\n",
       "0  b2897e1277f56641193a6db73825f707eed3e4c9   \n",
       "1  e3d0d482ebd9a8ba81c254cc433f314142e72174   \n",
       "2  00b1d99e70f779eb4ede50059db469c65e8c1469   \n",
       "3  cf584e00f637cbd8f1bb35f3f09f5ed07b71aeb0   \n",
       "4  dde02f11923815e6a16a31dd6298c46b109c5dfa   \n",
       "\n",
       "                                               title  \\\n",
       "0  Sequence requirements for RNA strand transfer ...   \n",
       "1  Crystal structure of murine sCEACAM1a[1,4]: a ...   \n",
       "2  Synthesis of a novel hepatitis C virus protein...   \n",
       "3  Structure of coronavirus main proteinase revea...   \n",
       "4  Discontinuous and non-discontinuous subgenomic...   \n",
       "\n",
       "                                            abstract                       doi  \n",
       "0  Nidovirus subgenomic mRNAs contain a leader se...  10.1093/emboj/20.24.7220  \n",
       "1  CEACAM1 is a member of the carcinoembryonic an...   10.1093/emboj/21.9.2076  \n",
       "2  Hepatitis C virus (HCV) is an important human ...  10.1093/emboj/20.14.3840  \n",
       "3  The key enzyme in coronavirus polyprotein proc...      10.1093/emboj/cdf327  \n",
       "4  Arteri-, corona-, toro- and roniviruses are ev...      10.1093/emboj/cdf635  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_covid = df_covid.drop(['language'], axis = 1) \n",
    "df_covid.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Spacy Parser and Tokenizer\n",
    "\n",
    "We will be using spacy for the pre-processing. We will use use en_core_sci_lg which is spacy's model for scientific and medical documents and create \"processed abstracts\" feature. It will later be used to calculate the centroid for each abstract."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Download the spacy bio parser\n",
    "with io.capture_output() as captured:\n",
    "    !pip install https://s3-us-west-2.amazonaws.com/ai2-s2-scispacy/releases/v0.2.4/en_core_sci_lg-0.2.4.tar.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "#NLP \n",
    "import en_core_sci_lg  # model downloaded in previous step"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "punctuations = string.punctuation\n",
    "\n",
    "stopwords = list(STOP_WORDS)\n",
    "custom_stop_words = [\n",
    "    'doi', 'preprint', 'copyright', 'peer', 'reviewed', 'org', 'https', 'et', 'al', 'author', 'figure', \n",
    "    'rights', 'reserved', 'permission', 'used', 'using', 'biorxiv', 'medrxiv', 'license', 'fig', 'fig.', \n",
    "    'al.', 'Elsevier', 'PMC', 'CZI', 'www'\n",
    "]\n",
    "\n",
    "for w in custom_stop_words:\n",
    "    if w not in stopwords:\n",
    "        stopwords.append(w)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "39d346bd287d4706829b16feec4bc090",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "HBox(children=(FloatProgress(value=0.0, max=38071.0), HTML(value='')))"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>paper_id</th>\n",
       "      <th>title</th>\n",
       "      <th>abstract</th>\n",
       "      <th>doi</th>\n",
       "      <th>processed_abstract</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>b2897e1277f56641193a6db73825f707eed3e4c9</td>\n",
       "      <td>Sequence requirements for RNA strand transfer ...</td>\n",
       "      <td>Nidovirus subgenomic mRNAs contain a leader se...</td>\n",
       "      <td>10.1093/emboj/20.24.7220</td>\n",
       "      <td>nidovirus subgenomic mrnas contain leader sequ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>e3d0d482ebd9a8ba81c254cc433f314142e72174</td>\n",
       "      <td>Crystal structure of murine sCEACAM1a[1,4]: a ...</td>\n",
       "      <td>CEACAM1 is a member of the carcinoembryonic an...</td>\n",
       "      <td>10.1093/emboj/21.9.2076</td>\n",
       "      <td>ceacam1 member carcinoembryonic antigen cea fa...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>00b1d99e70f779eb4ede50059db469c65e8c1469</td>\n",
       "      <td>Synthesis of a novel hepatitis C virus protein...</td>\n",
       "      <td>Hepatitis C virus (HCV) is an important human ...</td>\n",
       "      <td>10.1093/emboj/20.14.3840</td>\n",
       "      <td>hepatitis c virus hcv important human pathogen...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>cf584e00f637cbd8f1bb35f3f09f5ed07b71aeb0</td>\n",
       "      <td>Structure of coronavirus main proteinase revea...</td>\n",
       "      <td>The key enzyme in coronavirus polyprotein proc...</td>\n",
       "      <td>10.1093/emboj/cdf327</td>\n",
       "      <td>key enzyme coronavirus polyprotein process vir...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>dde02f11923815e6a16a31dd6298c46b109c5dfa</td>\n",
       "      <td>Discontinuous and non-discontinuous subgenomic...</td>\n",
       "      <td>Arteri-, corona-, toro- and roniviruses are ev...</td>\n",
       "      <td>10.1093/emboj/cdf635</td>\n",
       "      <td>arteri- corona- toro- roniviruses evolutionari...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                   paper_id  \\\n",
       "0  b2897e1277f56641193a6db73825f707eed3e4c9   \n",
       "1  e3d0d482ebd9a8ba81c254cc433f314142e72174   \n",
       "2  00b1d99e70f779eb4ede50059db469c65e8c1469   \n",
       "3  cf584e00f637cbd8f1bb35f3f09f5ed07b71aeb0   \n",
       "4  dde02f11923815e6a16a31dd6298c46b109c5dfa   \n",
       "\n",
       "                                               title  \\\n",
       "0  Sequence requirements for RNA strand transfer ...   \n",
       "1  Crystal structure of murine sCEACAM1a[1,4]: a ...   \n",
       "2  Synthesis of a novel hepatitis C virus protein...   \n",
       "3  Structure of coronavirus main proteinase revea...   \n",
       "4  Discontinuous and non-discontinuous subgenomic...   \n",
       "\n",
       "                                            abstract  \\\n",
       "0  Nidovirus subgenomic mRNAs contain a leader se...   \n",
       "1  CEACAM1 is a member of the carcinoembryonic an...   \n",
       "2  Hepatitis C virus (HCV) is an important human ...   \n",
       "3  The key enzyme in coronavirus polyprotein proc...   \n",
       "4  Arteri-, corona-, toro- and roniviruses are ev...   \n",
       "\n",
       "                        doi                                 processed_abstract  \n",
       "0  10.1093/emboj/20.24.7220  nidovirus subgenomic mrnas contain leader sequ...  \n",
       "1   10.1093/emboj/21.9.2076  ceacam1 member carcinoembryonic antigen cea fa...  \n",
       "2  10.1093/emboj/20.14.3840  hepatitis c virus hcv important human pathogen...  \n",
       "3      10.1093/emboj/cdf327  key enzyme coronavirus polyprotein process vir...  \n",
       "4      10.1093/emboj/cdf635  arteri- corona- toro- roniviruses evolutionari...  "
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Parser\n",
    "parser = en_core_sci_lg.load(disable=[\"tagger\", \"ner\"])\n",
    "parser.max_length = 7000000\n",
    "\n",
    "def spacy_tokenizer(sentence):\n",
    "    mytokens = parser(sentence)\n",
    "    mytokens = [ word.lemma_.lower().strip() if word.lemma_ != \"-PRON-\" else word.lower_ for word in mytokens ]\n",
    "    mytokens = [ word for word in mytokens if word not in stopwords and word not in punctuations ]\n",
    "    mytokens = \" \".join([i for i in mytokens])\n",
    "    return mytokens\n",
    "\n",
    "\n",
    "tqdm.pandas()\n",
    "df_covid[\"processed_abstract\"] = df_covid[\"abstract\"].progress_apply(spacy_tokenizer)\n",
    "df_covid.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Sentence Tokenization\n",
    "\n",
    "gensim's word2vec corpus should be in the form of separate sentences, therefore we will use spacy's tokenizer in order to split the corpus (all the abstracts) into sentences."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "6ef0e0ac4d574941b51a7d893e133c7f",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "HBox(children=(FloatProgress(value=0.0, max=38071.0), HTML(value='')))"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "#sentence tokenization to prepare the corpus\n",
    "abstracts = df_covid['abstract'].values\n",
    "\n",
    "nlp = en_core_sci_lg.load(disable = ['ner', 'tagger'])\n",
    "nlp.add_pipe(nlp.create_pipe('sentencizer'), before=\"parser\")\n",
    "word2vec_corpus = []\n",
    "\n",
    "for i in tqdm(range(0, len(abstracts))):\n",
    "    raw_text = abstracts[i]\n",
    "    doc = nlp(raw_text)\n",
    "    sentences = [sent.string.strip() for sent in doc.sents]\n",
    "    \n",
    "    for sent in sentences:\n",
    "        processed_sent = spacy_tokenizer(sent)\n",
    "        processed_sent_list = processed_sent.split(\" \")\n",
    "        word2vec_corpus.append(processed_sent_list)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "check out the corpus."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[['nidovirus',\n",
       "  'subgenomic',\n",
       "  'mrnas',\n",
       "  'contain',\n",
       "  'leader',\n",
       "  'sequence',\n",
       "  'derive',\n",
       "  '5′',\n",
       "  'end',\n",
       "  'genome',\n",
       "  'fuse',\n",
       "  'different',\n",
       "  'sequence',\n",
       "  '‘',\n",
       "  'body',\n",
       "  '’',\n",
       "  'derive',\n",
       "  '3′',\n",
       "  'end'],\n",
       " ['generation',\n",
       "  'involve',\n",
       "  'unique',\n",
       "  'mechanism',\n",
       "  'discontinuous',\n",
       "  'subgenomic',\n",
       "  'rna',\n",
       "  'synthesis',\n",
       "  'resemble',\n",
       "  'copy-choice',\n",
       "  'rna',\n",
       "  'recombination'],\n",
       " ['process',\n",
       "  'nascent',\n",
       "  'rna',\n",
       "  'strand',\n",
       "  'transfer',\n",
       "  'site',\n",
       "  'template',\n",
       "  'plus',\n",
       "  'minus',\n",
       "  'strand',\n",
       "  'synthesis',\n",
       "  'yield',\n",
       "  'subgenomic',\n",
       "  'rna',\n",
       "  'molecule'],\n",
       " ['central',\n",
       "  'process',\n",
       "  'transcription-regulating',\n",
       "  'sequence',\n",
       "  'trss',\n",
       "  'present',\n",
       "  'template',\n",
       "  'site',\n",
       "  'ensure',\n",
       "  'fidelity',\n",
       "  'strand',\n",
       "  'transfer'],\n",
       " ['present',\n",
       "  'result',\n",
       "  'comprehensive',\n",
       "  'co-variation',\n",
       "  'mutagenesis',\n",
       "  'study',\n",
       "  'equine',\n",
       "  'arteritis',\n",
       "  'virus',\n",
       "  'trss',\n",
       "  'demonstrate',\n",
       "  'discontinuous',\n",
       "  'rna',\n",
       "  'synthesis',\n",
       "  'depend',\n",
       "  'base',\n",
       "  'pair',\n",
       "  'sense',\n",
       "  'leader',\n",
       "  'trs',\n",
       "  'antisense',\n",
       "  'body',\n",
       "  'trs',\n",
       "  'primary',\n",
       "  'sequence',\n",
       "  'body',\n",
       "  'trs'],\n",
       " ['leader',\n",
       "  'trs',\n",
       "  'merely',\n",
       "  'play',\n",
       "  'target',\n",
       "  'role',\n",
       "  'strand',\n",
       "  'transfer',\n",
       "  'body',\n",
       "  'trs',\n",
       "  'fulfil',\n",
       "  'multiple',\n",
       "  'function'],\n",
       " ['sequence',\n",
       "  'mrna',\n",
       "  'leader',\n",
       "  '–',\n",
       "  'body',\n",
       "  'junction',\n",
       "  'trs',\n",
       "  'mutant',\n",
       "  'strongly',\n",
       "  'suggest',\n",
       "  'discontinuous',\n",
       "  'step',\n",
       "  'occur',\n",
       "  'minus',\n",
       "  'strand',\n",
       "  'synthesis'],\n",
       " ['ceacam1', 'member', 'carcinoembryonic', 'antigen', 'cea', 'family'],\n",
       " ['isoforms',\n",
       "  'murine',\n",
       "  'ceacam1',\n",
       "  'serve',\n",
       "  'receptor',\n",
       "  'mouse',\n",
       "  'hepatitis',\n",
       "  'virus',\n",
       "  'mhv',\n",
       "  'murine',\n",
       "  'coronavirus'],\n",
       " ['report',\n",
       "  'crystal',\n",
       "  'structure',\n",
       "  'soluble',\n",
       "  'murine',\n",
       "  'sceacam1a[1,4',\n",
       "  'compose',\n",
       "  'ig-like',\n",
       "  'domain',\n",
       "  'mhv',\n",
       "  'neutralize',\n",
       "  'activity']]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "word2vec_corpus[:10]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Word2vec Training\n",
    "\n",
    "We will use gensim's word2vec and train it on the corpus we prepared. `min_count` is the minimum count for a word to occur in the corpus in order to be mapped to a vector. `size` is the size of the vectors produced. `workers` is the number of cores. `window` is the context size to consider. `sg` is skipgram model. The `min_count`, `size` and `window` were calculated empirically. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Train the genisim word2vec model with our own custom corpus\n",
    "model = Word2Vec(word2vec_corpus, min_count=3,size= 50,workers=4, window =5, sg = 1)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Word Centroid Similarity (WCS)\n",
    "\n",
    "Now that we have word embedding for our corpus, the approach we will take is to measure the cosine similarity between the centroid of each document and the query. This idea is inspired from https://github.com/lgalke/vec4ir#word-centroid-similarity-wcs where it is mentioned:\n",
    "\n",
    ">An intuitive approach to use word embeddings in information retrieval is the word centroid similarity (WCS). The representation for each document is the centroid of its respective word vectors. Since word vectors carry semantic information of the words, one could assume that the centroid of the word vectors within a document encodes its meaning to some extent. At query time, the centroid of the query’s word vectors is computed. The cosine similarity to the centroids of the (matching) documents is used as a measure of relevance. When the initial word frequencies of the queries and documents are first re-weighted according to inverse-document frequency (i.e. frequent words in the corpus are discounted), the technique is labeled IDF re-weighted word centroid similarity (IWCS).\n",
    "\n",
    "\n",
    "For this purpose, we will calculate the centroid for each abstract using the vectors of all the words incorporating the anstract."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>paper_id</th>\n",
       "      <th>title</th>\n",
       "      <th>abstract</th>\n",
       "      <th>doi</th>\n",
       "      <th>processed_abstract</th>\n",
       "      <th>centroid</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>b2897e1277f56641193a6db73825f707eed3e4c9</td>\n",
       "      <td>Sequence requirements for RNA strand transfer ...</td>\n",
       "      <td>Nidovirus subgenomic mRNAs contain a leader se...</td>\n",
       "      <td>10.1093/emboj/20.24.7220</td>\n",
       "      <td>nidovirus subgenomic mrnas contain leader sequ...</td>\n",
       "      <td>[42.89778371725697, 17.383791841566563, -46.81...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>e3d0d482ebd9a8ba81c254cc433f314142e72174</td>\n",
       "      <td>Crystal structure of murine sCEACAM1a[1,4]: a ...</td>\n",
       "      <td>CEACAM1 is a member of the carcinoembryonic an...</td>\n",
       "      <td>10.1093/emboj/21.9.2076</td>\n",
       "      <td>ceacam1 member carcinoembryonic antigen cea fa...</td>\n",
       "      <td>[30.22426011785865, 13.17461103014648, -33.199...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>00b1d99e70f779eb4ede50059db469c65e8c1469</td>\n",
       "      <td>Synthesis of a novel hepatitis C virus protein...</td>\n",
       "      <td>Hepatitis C virus (HCV) is an important human ...</td>\n",
       "      <td>10.1093/emboj/20.14.3840</td>\n",
       "      <td>hepatitis c virus hcv important human pathogen...</td>\n",
       "      <td>[37.079803220462054, 12.635716843418777, -43.3...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>cf584e00f637cbd8f1bb35f3f09f5ed07b71aeb0</td>\n",
       "      <td>Structure of coronavirus main proteinase revea...</td>\n",
       "      <td>The key enzyme in coronavirus polyprotein proc...</td>\n",
       "      <td>10.1093/emboj/cdf327</td>\n",
       "      <td>key enzyme coronavirus polyprotein process vir...</td>\n",
       "      <td>[33.258001547306776, 11.860749502433464, -31.2...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>dde02f11923815e6a16a31dd6298c46b109c5dfa</td>\n",
       "      <td>Discontinuous and non-discontinuous subgenomic...</td>\n",
       "      <td>Arteri-, corona-, toro- and roniviruses are ev...</td>\n",
       "      <td>10.1093/emboj/cdf635</td>\n",
       "      <td>arteri- corona- toro- roniviruses evolutionari...</td>\n",
       "      <td>[47.382160579785705, 14.095707904547453, -42.6...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                   paper_id  \\\n",
       "0  b2897e1277f56641193a6db73825f707eed3e4c9   \n",
       "1  e3d0d482ebd9a8ba81c254cc433f314142e72174   \n",
       "2  00b1d99e70f779eb4ede50059db469c65e8c1469   \n",
       "3  cf584e00f637cbd8f1bb35f3f09f5ed07b71aeb0   \n",
       "4  dde02f11923815e6a16a31dd6298c46b109c5dfa   \n",
       "\n",
       "                                               title  \\\n",
       "0  Sequence requirements for RNA strand transfer ...   \n",
       "1  Crystal structure of murine sCEACAM1a[1,4]: a ...   \n",
       "2  Synthesis of a novel hepatitis C virus protein...   \n",
       "3  Structure of coronavirus main proteinase revea...   \n",
       "4  Discontinuous and non-discontinuous subgenomic...   \n",
       "\n",
       "                                            abstract  \\\n",
       "0  Nidovirus subgenomic mRNAs contain a leader se...   \n",
       "1  CEACAM1 is a member of the carcinoembryonic an...   \n",
       "2  Hepatitis C virus (HCV) is an important human ...   \n",
       "3  The key enzyme in coronavirus polyprotein proc...   \n",
       "4  Arteri-, corona-, toro- and roniviruses are ev...   \n",
       "\n",
       "                        doi  \\\n",
       "0  10.1093/emboj/20.24.7220   \n",
       "1   10.1093/emboj/21.9.2076   \n",
       "2  10.1093/emboj/20.14.3840   \n",
       "3      10.1093/emboj/cdf327   \n",
       "4      10.1093/emboj/cdf635   \n",
       "\n",
       "                                  processed_abstract  \\\n",
       "0  nidovirus subgenomic mrnas contain leader sequ...   \n",
       "1  ceacam1 member carcinoembryonic antigen cea fa...   \n",
       "2  hepatitis c virus hcv important human pathogen...   \n",
       "3  key enzyme coronavirus polyprotein process vir...   \n",
       "4  arteri- corona- toro- roniviruses evolutionari...   \n",
       "\n",
       "                                            centroid  \n",
       "0  [42.89778371725697, 17.383791841566563, -46.81...  \n",
       "1  [30.22426011785865, 13.17461103014648, -33.199...  \n",
       "2  [37.079803220462054, 12.635716843418777, -43.3...  \n",
       "3  [33.258001547306776, 11.860749502433464, -31.2...  \n",
       "4  [47.382160579785705, 14.095707904547453, -42.6...  "
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#calculate the centroid for each abstract\n",
    "\n",
    "a = [0.0]*50\n",
    "df_covid[\"centroid\"] = [a]*df_covid.shape[0]\n",
    "\n",
    "for index, row in df_covid.iterrows():\n",
    "    abstract = row['processed_abstract']\n",
    "    total_sim = 0\n",
    "    words = abstract.split(\" \")\n",
    "    centroid = np.array([0.0]*50)\n",
    "    for word in words:\n",
    "        try:\n",
    "            b = model[word]\n",
    "        except:\n",
    "            continue\n",
    "        centroid = np.add(centroid, b)\n",
    "\n",
    "    df_covid.at[index,'centroid'] = centroid.tolist()\n",
    "\n",
    "df_covid.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Ranking documents\n",
    "\n",
    "Now we will create a function that given a query, would rank the documents from most similar to least similar."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "def rank_docs(model, query, df_covid, num) :\n",
    "    #[(paper_id, processed_abstract, url, cosine_sim)]\n",
    "    cosine_list = []\n",
    "    \n",
    "    a = []\n",
    "    query = query.split(\" \")\n",
    "    for q in query:\n",
    "        try:\n",
    "            a.append(model[q])\n",
    "        except:\n",
    "            continue\n",
    "    \n",
    "    for index, row in df_covid.iterrows():\n",
    "        centroid = row['centroid']\n",
    "        total_sim = 0\n",
    "        for a_i in a:\n",
    "            cos_sim = np.dot(a_i, centroid)/(np.linalg.norm(a_i)*np.linalg.norm(centroid))\n",
    "            total_sim += cos_sim\n",
    "        cosine_list.append((row['title'], row['doi'], total_sim)) \n",
    "    \n",
    "    \n",
    "    cosine_list.sort(key=lambda x:x[2], reverse=True) ## in Descedning order \n",
    "    \n",
    "    papers_list = []\n",
    "    for item in cosine_list[:num]:\n",
    "        papers_list.append((item[0], item[1], item[2]))\n",
    "    return papers_list"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Saving the model and the dataframe"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## saving the model and the data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "model.save(\"./model.model\")\n",
    "df_covid.to_pickle(\"./df_covid.pkl\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Loading the model and the data\n",
    "The save/load steps are done in order to avoid re-training the model each time."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "saved_model = Word2Vec.load(\"./model.model\")\n",
    "saved_df_covid = pd.read_pickle(\"./df_covid.pkl\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Results\n",
    "\n",
    "Now let's see some results using our model. query() function takes the query string as an input, together with a number representing the top matches you want, and print the titles of the top matches most relevant articles retrieved for each query, clicking the title a new tab will open with the paper."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "def query(the_query, top_matches=10):\n",
    "    q = spacy_tokenizer(the_query)\n",
    "    try:\n",
    "        model_to_use = model\n",
    "    except:\n",
    "        model_to_use = saved_model\n",
    "    try:\n",
    "        df_covid_to_use = df_covid\n",
    "    except:\n",
    "        df_covid_to_use = saved_df_covid\n",
    "    results = rank_docs(model_to_use, q, df_covid_to_use, top_matches)\n",
    "    html = \"\"\"\n",
    "    <html>\n",
    "    <style>\n",
    "        body {\n",
    "            margin: 0;\n",
    "            padding: 0;\n",
    "            background: #ffffff;\n",
    "            font-family: sans-serif;\n",
    "        }\n",
    "\n",
    "        ul {\n",
    "            position: relative;\n",
    "            width: 100%;\n",
    "            margin: 100px auto 0;\n",
    "            padding: 20px;\n",
    "            box-sizing: border-box;\n",
    "            background: rgba(0, 0, 0, 0);\n",
    "            box-shadow: inset 0 0 15px rgba(0, 0, 0, .2);\n",
    "            border-radius: 5px;\n",
    "            overflow: hidden;\n",
    "        }\n",
    "\n",
    "        ul li {\n",
    "            display: flex;\n",
    "            background: rgba(255, 0, 0, 0.25);\n",
    "            padding: 10px 20px;\n",
    "            color: #fff;\n",
    "            margin: 5px 0;\n",
    "            transition: .6s;\n",
    "            border-radius: 5px;\n",
    "        }\n",
    "\n",
    "        ul li:hover {\n",
    "            transform: scale(1.02);\n",
    "            background: rgba(255, 0, 0, 0.5);\n",
    "        }\n",
    "    </style>\n",
    "        \n",
    "        <body>\n",
    "        <ul>\n",
    "    \"\"\"\n",
    "    for i in range(len(results)):\n",
    "        paper_name = results[i][0]\n",
    "        paper_doi = results[i][1]\n",
    "        paper_link = \"https://doi.org/\" + str(paper_doi)\n",
    "        html += \"\"\"\n",
    "                <li>\n",
    "                <a href=\" \"\"\" + str(paper_link) + \"\"\" \" target=\"_blank\">\n",
    "                    <div>\n",
    "                        <h3>\n",
    "                            \"\"\" + str(i+1) + \"&emsp;\" +str(paper_name) + \"\"\" \n",
    "                        </h3>\n",
    "                        <cite>\n",
    "                            doi\n",
    "                            <span >\n",
    "                                > \"\"\" + str(paper_doi) + \"\"\"\n",
    "                            </span>\n",
    "                        </cite>\n",
    "                    </div>\n",
    "                </a>\n",
    "        \"\"\"\n",
    "        \n",
    "    html += \"</body></html>\"\n",
    "    display(HTML(html))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "> # What do we know about virus genetics, origin, and evolution?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "    <html>\n",
       "    <style>\n",
       "        body {\n",
       "            margin: 0;\n",
       "            padding: 0;\n",
       "            background: #ffffff;\n",
       "            font-family: sans-serif;\n",
       "        }\n",
       "\n",
       "        ul {\n",
       "            position: relative;\n",
       "            width: 100%;\n",
       "            margin: 100px auto 0;\n",
       "            padding: 20px;\n",
       "            box-sizing: border-box;\n",
       "            background: rgba(0, 0, 0, 0);\n",
       "            box-shadow: inset 0 0 15px rgba(0, 0, 0, .2);\n",
       "            border-radius: 5px;\n",
       "            overflow: hidden;\n",
       "        }\n",
       "\n",
       "        ul li {\n",
       "            display: flex;\n",
       "            background: rgba(255, 0, 0, 0.25);\n",
       "            padding: 10px 20px;\n",
       "            color: #fff;\n",
       "            margin: 5px 0;\n",
       "            transition: .6s;\n",
       "            border-radius: 5px;\n",
       "        }\n",
       "\n",
       "        ul li:hover {\n",
       "            transform: scale(1.02);\n",
       "            background: rgba(255, 0, 0, 0.5);\n",
       "        }\n",
       "    </style>\n",
       "        \n",
       "        <body>\n",
       "        <ul>\n",
       "    \n",
       "                <li>\n",
       "                <a href=\" https://doi.org/10.1016/j.meegid.2020.104285 \" target=\"_blank\">\n",
       "                    <div>\n",
       "                        <h3>\n",
       "                            1&emsp;Genetic evolution analysis of 2019 novel coronavirus and coronavirus from other species \n",
       "                        </h3>\n",
       "                        <cite>\n",
       "                            doi\n",
       "                            <span >\n",
       "                                > 10.1016/j.meegid.2020.104285\n",
       "                            </span>\n",
       "                        </cite>\n",
       "                    </div>\n",
       "                </a>\n",
       "        \n",
       "                <li>\n",
       "                <a href=\" https://doi.org/10.3201/eid1212.060401 \" target=\"_blank\">\n",
       "                    <div>\n",
       "                        <h3>\n",
       "                            2&emsp;Review of Bats and SARS \n",
       "                        </h3>\n",
       "                        <cite>\n",
       "                            doi\n",
       "                            <span >\n",
       "                                > 10.3201/eid1212.060401\n",
       "                            </span>\n",
       "                        </cite>\n",
       "                    </div>\n",
       "                </a>\n",
       "        \n",
       "                <li>\n",
       "                <a href=\" https://doi.org/10.1186/s12985-015-0422-1 \" target=\"_blank\">\n",
       "                    <div>\n",
       "                        <h3>\n",
       "                            3&emsp;Bat origin of human coronaviruses \n",
       "                        </h3>\n",
       "                        <cite>\n",
       "                            doi\n",
       "                            <span >\n",
       "                                > 10.1186/s12985-015-0422-1\n",
       "                            </span>\n",
       "                        </cite>\n",
       "                    </div>\n",
       "                </a>\n",
       "        \n",
       "                <li>\n",
       "                <a href=\" https://doi.org/10.1038/s41579-018-0118-9 \" target=\"_blank\">\n",
       "                    <div>\n",
       "                        <h3>\n",
       "                            4&emsp;Origin and evolution of pathogenic coronaviruses \n",
       "                        </h3>\n",
       "                        <cite>\n",
       "                            doi\n",
       "                            <span >\n",
       "                                > 10.1038/s41579-018-0118-9\n",
       "                            </span>\n",
       "                        </cite>\n",
       "                    </div>\n",
       "                </a>\n",
       "        \n",
       "                <li>\n",
       "                <a href=\" https://doi.org/10.1007/978-3-642-40605-8_13 \" target=\"_blank\">\n",
       "                    <div>\n",
       "                        <h3>\n",
       "                            5&emsp;Coronavirus \n",
       "                        </h3>\n",
       "                        <cite>\n",
       "                            doi\n",
       "                            <span >\n",
       "                                > 10.1007/978-3-642-40605-8_13\n",
       "                            </span>\n",
       "                        </cite>\n",
       "                    </div>\n",
       "                </a>\n",
       "        \n",
       "                <li>\n",
       "                <a href=\" https://doi.org/10.1007/978-3-540-39026-8_217 \" target=\"_blank\">\n",
       "                    <div>\n",
       "                        <h3>\n",
       "                            6&emsp;Coronavirus, humanpathogenes \n",
       "                        </h3>\n",
       "                        <cite>\n",
       "                            doi\n",
       "                            <span >\n",
       "                                > 10.1007/978-3-540-39026-8_217\n",
       "                            </span>\n",
       "                        </cite>\n",
       "                    </div>\n",
       "                </a>\n",
       "        \n",
       "                <li>\n",
       "                <a href=\" https://doi.org/10.1007/s11427-016-5039-0 \" target=\"_blank\">\n",
       "                    <div>\n",
       "                        <h3>\n",
       "                            7&emsp;Genetic diversity of coronaviruses in Miniopterus fuliginosus bats \n",
       "                        </h3>\n",
       "                        <cite>\n",
       "                            doi\n",
       "                            <span >\n",
       "                                > 10.1007/s11427-016-5039-0\n",
       "                            </span>\n",
       "                        </cite>\n",
       "                    </div>\n",
       "                </a>\n",
       "        \n",
       "                <li>\n",
       "                <a href=\" https://doi.org/10.3390/v4123689 \" target=\"_blank\">\n",
       "                    <div>\n",
       "                        <h3>\n",
       "                            8&emsp;Identification and Characterization of a Novel Alpaca Respiratory Coronavirus Most Closely Related to the Human Coronavirus 229E \n",
       "                        </h3>\n",
       "                        <cite>\n",
       "                            doi\n",
       "                            <span >\n",
       "                                > 10.3390/v4123689\n",
       "                            </span>\n",
       "                        </cite>\n",
       "                    </div>\n",
       "                </a>\n",
       "        \n",
       "                <li>\n",
       "                <a href=\" https://doi.org/10.1016/j.meegid.2019.02.001 \" target=\"_blank\">\n",
       "                    <div>\n",
       "                        <h3>\n",
       "                            9&emsp;Geographical structure of bat SARS-related coronaviruses \n",
       "                        </h3>\n",
       "                        <cite>\n",
       "                            doi\n",
       "                            <span >\n",
       "                                > 10.1016/j.meegid.2019.02.001\n",
       "                            </span>\n",
       "                        </cite>\n",
       "                    </div>\n",
       "                </a>\n",
       "        \n",
       "                <li>\n",
       "                <a href=\" https://doi.org/10.1007/s00705-010-0612-5 \" target=\"_blank\">\n",
       "                    <div>\n",
       "                        <h3>\n",
       "                            10&emsp;Identification of SARS-like coronaviruses in horseshoe bats (Rhinolophus hipposideros) in Slovenia \n",
       "                        </h3>\n",
       "                        <cite>\n",
       "                            doi\n",
       "                            <span >\n",
       "                                > 10.1007/s00705-010-0612-5\n",
       "                            </span>\n",
       "                        </cite>\n",
       "                    </div>\n",
       "                </a>\n",
       "        </body></html>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "query('origin of coronavirus')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "    <html>\n",
       "    <style>\n",
       "        body {\n",
       "            margin: 0;\n",
       "            padding: 0;\n",
       "            background: #ffffff;\n",
       "            font-family: sans-serif;\n",
       "        }\n",
       "\n",
       "        ul {\n",
       "            position: relative;\n",
       "            width: 100%;\n",
       "            margin: 100px auto 0;\n",
       "            padding: 20px;\n",
       "            box-sizing: border-box;\n",
       "            background: rgba(0, 0, 0, 0);\n",
       "            box-shadow: inset 0 0 15px rgba(0, 0, 0, .2);\n",
       "            border-radius: 5px;\n",
       "            overflow: hidden;\n",
       "        }\n",
       "\n",
       "        ul li {\n",
       "            display: flex;\n",
       "            background: rgba(255, 0, 0, 0.25);\n",
       "            padding: 10px 20px;\n",
       "            color: #fff;\n",
       "            margin: 5px 0;\n",
       "            transition: .6s;\n",
       "            border-radius: 5px;\n",
       "        }\n",
       "\n",
       "        ul li:hover {\n",
       "            transform: scale(1.02);\n",
       "            background: rgba(255, 0, 0, 0.5);\n",
       "        }\n",
       "    </style>\n",
       "        \n",
       "        <body>\n",
       "        <ul>\n",
       "    \n",
       "                <li>\n",
       "                <a href=\" https://doi.org/10.1101/2020.04.22.055863 \" target=\"_blank\">\n",
       "                    <div>\n",
       "                        <h3>\n",
       "                            1&emsp;Genomic, geographic and temporal distributions of SARS-CoV-2 mutations \n",
       "                        </h3>\n",
       "                        <cite>\n",
       "                            doi\n",
       "                            <span >\n",
       "                                > 10.1101/2020.04.22.055863\n",
       "                            </span>\n",
       "                        </cite>\n",
       "                    </div>\n",
       "                </a>\n",
       "        \n",
       "                <li>\n",
       "                <a href=\" https://doi.org/10.1101/2020.03.30.20047878 \" target=\"_blank\">\n",
       "                    <div>\n",
       "                        <h3>\n",
       "                            2&emsp;ACE2 and TMPRSS2 variants and expression as candidates to sex and country differences in COVID-19 severity in Italy \n",
       "                        </h3>\n",
       "                        <cite>\n",
       "                            doi\n",
       "                            <span >\n",
       "                                > 10.1101/2020.03.30.20047878\n",
       "                            </span>\n",
       "                        </cite>\n",
       "                    </div>\n",
       "                </a>\n",
       "        \n",
       "                <li>\n",
       "                <a href=\" https://doi.org/10.1111/jvh.13006 \" target=\"_blank\">\n",
       "                    <div>\n",
       "                        <h3>\n",
       "                            3&emsp;Origin and dissemination of hepatitis B virus genotype C in East Asia revealed by phylodynamic analysis and historical correlates \n",
       "                        </h3>\n",
       "                        <cite>\n",
       "                            doi\n",
       "                            <span >\n",
       "                                > 10.1111/jvh.13006\n",
       "                            </span>\n",
       "                        </cite>\n",
       "                    </div>\n",
       "                </a>\n",
       "        \n",
       "                <li>\n",
       "                <a href=\" https://doi.org/10.1016/j.genrep.2020.100682 \" target=\"_blank\">\n",
       "                    <div>\n",
       "                        <h3>\n",
       "                            4&emsp;Genomic characterization of a novel SARS-CoV-2 \n",
       "                        </h3>\n",
       "                        <cite>\n",
       "                            doi\n",
       "                            <span >\n",
       "                                > 10.1016/j.genrep.2020.100682\n",
       "                            </span>\n",
       "                        </cite>\n",
       "                    </div>\n",
       "                </a>\n",
       "        \n",
       "                <li>\n",
       "                <a href=\" https://doi.org/10.1186/s12985-015-0444-8 \" target=\"_blank\">\n",
       "                    <div>\n",
       "                        <h3>\n",
       "                            5&emsp;Tackling dengue fever: Current status and challenges \n",
       "                        </h3>\n",
       "                        <cite>\n",
       "                            doi\n",
       "                            <span >\n",
       "                                > 10.1186/s12985-015-0444-8\n",
       "                            </span>\n",
       "                        </cite>\n",
       "                    </div>\n",
       "                </a>\n",
       "        \n",
       "                <li>\n",
       "                <a href=\" https://doi.org/10.1016/s1473-3099(06)70601-6 \" target=\"_blank\">\n",
       "                    <div>\n",
       "                        <h3>\n",
       "                            6&emsp;Genetic susceptibility to infectious diseases: big is beautiful, but will bigger be even better? \n",
       "                        </h3>\n",
       "                        <cite>\n",
       "                            doi\n",
       "                            <span >\n",
       "                                > 10.1016/s1473-3099(06)70601-6\n",
       "                            </span>\n",
       "                        </cite>\n",
       "                    </div>\n",
       "                </a>\n",
       "        \n",
       "                <li>\n",
       "                <a href=\" https://doi.org/10.1016/j.meegid.2020.104260 \" target=\"_blank\">\n",
       "                    <div>\n",
       "                        <h3>\n",
       "                            7&emsp;Genetic diversity and evolution of SARS-CoV-2 \n",
       "                        </h3>\n",
       "                        <cite>\n",
       "                            doi\n",
       "                            <span >\n",
       "                                > 10.1016/j.meegid.2020.104260\n",
       "                            </span>\n",
       "                        </cite>\n",
       "                    </div>\n",
       "                </a>\n",
       "        </body></html>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "query('covid19 genetics', top_matches=7)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "> # What is known about transmission, incubation, and environmental stability?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "    <html>\n",
       "    <style>\n",
       "        body {\n",
       "            margin: 0;\n",
       "            padding: 0;\n",
       "            background: #ffffff;\n",
       "            font-family: sans-serif;\n",
       "        }\n",
       "\n",
       "        ul {\n",
       "            position: relative;\n",
       "            width: 100%;\n",
       "            margin: 100px auto 0;\n",
       "            padding: 20px;\n",
       "            box-sizing: border-box;\n",
       "            background: rgba(0, 0, 0, 0);\n",
       "            box-shadow: inset 0 0 15px rgba(0, 0, 0, .2);\n",
       "            border-radius: 5px;\n",
       "            overflow: hidden;\n",
       "        }\n",
       "\n",
       "        ul li {\n",
       "            display: flex;\n",
       "            background: rgba(255, 0, 0, 0.25);\n",
       "            padding: 10px 20px;\n",
       "            color: #fff;\n",
       "            margin: 5px 0;\n",
       "            transition: .6s;\n",
       "            border-radius: 5px;\n",
       "        }\n",
       "\n",
       "        ul li:hover {\n",
       "            transform: scale(1.02);\n",
       "            background: rgba(255, 0, 0, 0.5);\n",
       "        }\n",
       "    </style>\n",
       "        \n",
       "        <body>\n",
       "        <ul>\n",
       "    \n",
       "                <li>\n",
       "                <a href=\" https://doi.org/10.1101/2020.02.08.20021311 \" target=\"_blank\">\n",
       "                    <div>\n",
       "                        <h3>\n",
       "                            1&emsp;Assessing the plausibility of subcritical transmission of 2019-nCoV in the United States \n",
       "                        </h3>\n",
       "                        <cite>\n",
       "                            doi\n",
       "                            <span >\n",
       "                                > 10.1101/2020.02.08.20021311\n",
       "                            </span>\n",
       "                        </cite>\n",
       "                    </div>\n",
       "                </a>\n",
       "        \n",
       "                <li>\n",
       "                <a href=\" https://doi.org/10.1111/ina.12430 \" target=\"_blank\">\n",
       "                    <div>\n",
       "                        <h3>\n",
       "                            2&emsp;A study of the probable transmission routes of MERS‐CoV during the first hospital outbreak in the Republic of Korea \n",
       "                        </h3>\n",
       "                        <cite>\n",
       "                            doi\n",
       "                            <span >\n",
       "                                > 10.1111/ina.12430\n",
       "                            </span>\n",
       "                        </cite>\n",
       "                    </div>\n",
       "                </a>\n",
       "        \n",
       "                <li>\n",
       "                <a href=\" https://doi.org/10.1101/2020.04.06.20055228 \" target=\"_blank\">\n",
       "                    <div>\n",
       "                        <h3>\n",
       "                            3&emsp;Multi-route respiratory infection: when a transmission route may dominate \n",
       "                        </h3>\n",
       "                        <cite>\n",
       "                            doi\n",
       "                            <span >\n",
       "                                > 10.1101/2020.04.06.20055228\n",
       "                            </span>\n",
       "                        </cite>\n",
       "                    </div>\n",
       "                </a>\n",
       "        \n",
       "                <li>\n",
       "                <a href=\" https://doi.org/10.1016/j.mehy.2020.109781 \" target=\"_blank\">\n",
       "                    <div>\n",
       "                        <h3>\n",
       "                            4&emsp;Airborne route and bad use of ventilation systems as non-negligible factors in SARS-CoV-2 transmission \n",
       "                        </h3>\n",
       "                        <cite>\n",
       "                            doi\n",
       "                            <span >\n",
       "                                > 10.1016/j.mehy.2020.109781\n",
       "                            </span>\n",
       "                        </cite>\n",
       "                    </div>\n",
       "                </a>\n",
       "        \n",
       "                <li>\n",
       "                <a href=\" https://doi.org/10.1371/journal.pone.0181558 \" target=\"_blank\">\n",
       "                    <div>\n",
       "                        <h3>\n",
       "                            5&emsp;Role of fomites in SARS transmission during the largest hospital outbreak in Hong Kong \n",
       "                        </h3>\n",
       "                        <cite>\n",
       "                            doi\n",
       "                            <span >\n",
       "                                > 10.1371/journal.pone.0181558\n",
       "                            </span>\n",
       "                        </cite>\n",
       "                    </div>\n",
       "                </a>\n",
       "        \n",
       "                <li>\n",
       "                <a href=\" https://doi.org/10.1016/j.coviro.2016.11.001 \" target=\"_blank\">\n",
       "                    <div>\n",
       "                        <h3>\n",
       "                            6&emsp;Sustained fecal-oral human-to-human transmission following a zoonotic event \n",
       "                        </h3>\n",
       "                        <cite>\n",
       "                            doi\n",
       "                            <span >\n",
       "                                > 10.1016/j.coviro.2016.11.001\n",
       "                            </span>\n",
       "                        </cite>\n",
       "                    </div>\n",
       "                </a>\n",
       "        \n",
       "                <li>\n",
       "                <a href=\" https://doi.org/10.1016/j.jtbi.2018.08.017 \" target=\"_blank\">\n",
       "                    <div>\n",
       "                        <h3>\n",
       "                            7&emsp;Stochastic dynamics of an epidemic with recurrent spillovers from an endemic reservoir \n",
       "                        </h3>\n",
       "                        <cite>\n",
       "                            doi\n",
       "                            <span >\n",
       "                                > 10.1016/j.jtbi.2018.08.017\n",
       "                            </span>\n",
       "                        </cite>\n",
       "                    </div>\n",
       "                </a>\n",
       "        \n",
       "                <li>\n",
       "                <a href=\" https://doi.org/10.1038/s41368-020-0075-9 \" target=\"_blank\">\n",
       "                    <div>\n",
       "                        <h3>\n",
       "                            8&emsp;Transmission routes of 2019-nCoV and controls in dental practice \n",
       "                        </h3>\n",
       "                        <cite>\n",
       "                            doi\n",
       "                            <span >\n",
       "                                > 10.1038/s41368-020-0075-9\n",
       "                            </span>\n",
       "                        </cite>\n",
       "                    </div>\n",
       "                </a>\n",
       "        \n",
       "                <li>\n",
       "                <a href=\" https://doi.org/10.1371/journal.pntd.0003398 \" target=\"_blank\">\n",
       "                    <div>\n",
       "                        <h3>\n",
       "                            9&emsp;Using Modelling to Disentangle the Relative Contributions of Zoonotic and Anthroponotic Transmission: The Case of Lassa Fever \n",
       "                        </h3>\n",
       "                        <cite>\n",
       "                            doi\n",
       "                            <span >\n",
       "                                > 10.1371/journal.pntd.0003398\n",
       "                            </span>\n",
       "                        </cite>\n",
       "                    </div>\n",
       "                </a>\n",
       "        \n",
       "                <li>\n",
       "                <a href=\" https://doi.org/10.1016/j.coviro.2016.11.004 \" target=\"_blank\">\n",
       "                    <div>\n",
       "                        <h3>\n",
       "                            10&emsp;Factors determining human-to-human transmissibility of zoonotic pathogens via contact \n",
       "                        </h3>\n",
       "                        <cite>\n",
       "                            doi\n",
       "                            <span >\n",
       "                                > 10.1016/j.coviro.2016.11.004\n",
       "                            </span>\n",
       "                        </cite>\n",
       "                    </div>\n",
       "                </a>\n",
       "        </body></html>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "query('transmission')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "    <html>\n",
       "    <style>\n",
       "        body {\n",
       "            margin: 0;\n",
       "            padding: 0;\n",
       "            background: #ffffff;\n",
       "            font-family: sans-serif;\n",
       "        }\n",
       "\n",
       "        ul {\n",
       "            position: relative;\n",
       "            width: 100%;\n",
       "            margin: 100px auto 0;\n",
       "            padding: 20px;\n",
       "            box-sizing: border-box;\n",
       "            background: rgba(0, 0, 0, 0);\n",
       "            box-shadow: inset 0 0 15px rgba(0, 0, 0, .2);\n",
       "            border-radius: 5px;\n",
       "            overflow: hidden;\n",
       "        }\n",
       "\n",
       "        ul li {\n",
       "            display: flex;\n",
       "            background: rgba(255, 0, 0, 0.25);\n",
       "            padding: 10px 20px;\n",
       "            color: #fff;\n",
       "            margin: 5px 0;\n",
       "            transition: .6s;\n",
       "            border-radius: 5px;\n",
       "        }\n",
       "\n",
       "        ul li:hover {\n",
       "            transform: scale(1.02);\n",
       "            background: rgba(255, 0, 0, 0.5);\n",
       "        }\n",
       "    </style>\n",
       "        \n",
       "        <body>\n",
       "        <ul>\n",
       "    \n",
       "                <li>\n",
       "                <a href=\" https://doi.org/10.1017/ice.2020.81 \" target=\"_blank\">\n",
       "                    <div>\n",
       "                        <h3>\n",
       "                            1&emsp;The difference in the incubation period of 2019 novel coronavirus (SARS-CoV-2) infection between travelers to Hubei and non-travelers: The need of a longer quarantine period \n",
       "                        </h3>\n",
       "                        <cite>\n",
       "                            doi\n",
       "                            <span >\n",
       "                                > 10.1017/ice.2020.81\n",
       "                            </span>\n",
       "                        </cite>\n",
       "                    </div>\n",
       "                </a>\n",
       "        \n",
       "                <li>\n",
       "                <a href=\" https://doi.org/10.3961/jpmph.20.048 \" target=\"_blank\">\n",
       "                    <div>\n",
       "                        <h3>\n",
       "                            2&emsp;A Chinese Case of Coronavirus Disease 2019 (COVID-19) Did Not Show Infectivity During the Incubation Period: Based on an Epidemiological Survey \n",
       "                        </h3>\n",
       "                        <cite>\n",
       "                            doi\n",
       "                            <span >\n",
       "                                > 10.3961/jpmph.20.048\n",
       "                            </span>\n",
       "                        </cite>\n",
       "                    </div>\n",
       "                </a>\n",
       "        \n",
       "                <li>\n",
       "                <a href=\" https://doi.org/10.1101/2020.04.02.20051318 \" target=\"_blank\">\n",
       "                    <div>\n",
       "                        <h3>\n",
       "                            3&emsp;Estimating the presymptomatic transmission of COVID19 using incubation period and serial interval data \n",
       "                        </h3>\n",
       "                        <cite>\n",
       "                            doi\n",
       "                            <span >\n",
       "                                > 10.1101/2020.04.02.20051318\n",
       "                            </span>\n",
       "                        </cite>\n",
       "                    </div>\n",
       "                </a>\n",
       "        \n",
       "                <li>\n",
       "                <a href=\" https://doi.org/10.3390/jcm9020538 \" target=\"_blank\">\n",
       "                    <div>\n",
       "                        <h3>\n",
       "                            4&emsp;Incubation Period and Other Epidemiological Characteristics of 2019 Novel Coronavirus Infections with Right Truncation: A Statistical Analysis of Publicly Available Case Data \n",
       "                        </h3>\n",
       "                        <cite>\n",
       "                            doi\n",
       "                            <span >\n",
       "                                > 10.3390/jcm9020538\n",
       "                            </span>\n",
       "                        </cite>\n",
       "                    </div>\n",
       "                </a>\n",
       "        \n",
       "                <li>\n",
       "                <a href=\" https://doi.org/10.1101/2020.02.26.20028431 \" target=\"_blank\">\n",
       "                    <div>\n",
       "                        <h3>\n",
       "                            5&emsp;Transmission characteristics of the COVID-19 outbreak in China: a study driven by data \n",
       "                        </h3>\n",
       "                        <cite>\n",
       "                            doi\n",
       "                            <span >\n",
       "                                > 10.1101/2020.02.26.20028431\n",
       "                            </span>\n",
       "                        </cite>\n",
       "                    </div>\n",
       "                </a>\n",
       "        \n",
       "                <li>\n",
       "                <a href=\" https://doi.org/10.1101/2020.01.26.20018754 \" target=\"_blank\">\n",
       "                    <div>\n",
       "                        <h3>\n",
       "                            6&emsp;Incubation Period and Other Epidemiological Characteristics of 2019 Novel Coronavirus Infections with Right Truncation: A Statistical Analysis of Publicly Available Case Data \n",
       "                        </h3>\n",
       "                        <cite>\n",
       "                            doi\n",
       "                            <span >\n",
       "                                > 10.1101/2020.01.26.20018754\n",
       "                            </span>\n",
       "                        </cite>\n",
       "                    </div>\n",
       "                </a>\n",
       "        \n",
       "                <li>\n",
       "                <a href=\" https://doi.org/10.1101/2020.03.06.20031955 \" target=\"_blank\">\n",
       "                    <div>\n",
       "                        <h3>\n",
       "                            7&emsp;Transmission of corona virus disease 2019 during the incubation period may lead to a quarantine loophole \n",
       "                        </h3>\n",
       "                        <cite>\n",
       "                            doi\n",
       "                            <span >\n",
       "                                > 10.1101/2020.03.06.20031955\n",
       "                            </span>\n",
       "                        </cite>\n",
       "                    </div>\n",
       "                </a>\n",
       "        \n",
       "                <li>\n",
       "                <a href=\" https://doi.org/10.1101/2020.03.03.20029983 \" target=\"_blank\">\n",
       "                    <div>\n",
       "                        <h3>\n",
       "                            8&emsp;Transmission interval estimates suggest pre-symptomatic spread of COVID-19 \n",
       "                        </h3>\n",
       "                        <cite>\n",
       "                            doi\n",
       "                            <span >\n",
       "                                > 10.1101/2020.03.03.20029983\n",
       "                            </span>\n",
       "                        </cite>\n",
       "                    </div>\n",
       "                </a>\n",
       "        \n",
       "                <li>\n",
       "                <a href=\" https://doi.org/10.1007/s11538-005-9005-4 \" target=\"_blank\">\n",
       "                    <div>\n",
       "                        <h3>\n",
       "                            9&emsp;Bayesian modelling of an epidemic of severe acute respiratory syndrome \n",
       "                        </h3>\n",
       "                        <cite>\n",
       "                            doi\n",
       "                            <span >\n",
       "                                > 10.1007/s11538-005-9005-4\n",
       "                            </span>\n",
       "                        </cite>\n",
       "                    </div>\n",
       "                </a>\n",
       "        \n",
       "                <li>\n",
       "                <a href=\" https://doi.org/10.1101/2020.03.21.20040329 \" target=\"_blank\">\n",
       "                    <div>\n",
       "                        <h3>\n",
       "                            10&emsp;Epidemiological parameters of coronavirus disease 2019: a pooled analysis of publicly reported individual data of 1155 cases from seven countries \n",
       "                        </h3>\n",
       "                        <cite>\n",
       "                            doi\n",
       "                            <span >\n",
       "                                > 10.1101/2020.03.21.20040329\n",
       "                            </span>\n",
       "                        </cite>\n",
       "                    </div>\n",
       "                </a>\n",
       "        </body></html>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "query('incubation period of covid19')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "    <html>\n",
       "    <style>\n",
       "        body {\n",
       "            margin: 0;\n",
       "            padding: 0;\n",
       "            background: #ffffff;\n",
       "            font-family: sans-serif;\n",
       "        }\n",
       "\n",
       "        ul {\n",
       "            position: relative;\n",
       "            width: 100%;\n",
       "            margin: 100px auto 0;\n",
       "            padding: 20px;\n",
       "            box-sizing: border-box;\n",
       "            background: rgba(0, 0, 0, 0);\n",
       "            box-shadow: inset 0 0 15px rgba(0, 0, 0, .2);\n",
       "            border-radius: 5px;\n",
       "            overflow: hidden;\n",
       "        }\n",
       "\n",
       "        ul li {\n",
       "            display: flex;\n",
       "            background: rgba(255, 0, 0, 0.25);\n",
       "            padding: 10px 20px;\n",
       "            color: #fff;\n",
       "            margin: 5px 0;\n",
       "            transition: .6s;\n",
       "            border-radius: 5px;\n",
       "        }\n",
       "\n",
       "        ul li:hover {\n",
       "            transform: scale(1.02);\n",
       "            background: rgba(255, 0, 0, 0.5);\n",
       "        }\n",
       "    </style>\n",
       "        \n",
       "        <body>\n",
       "        <ul>\n",
       "    \n",
       "                <li>\n",
       "                <a href=\" https://doi.org/10.1093/ve/vez045 \" target=\"_blank\">\n",
       "                    <div>\n",
       "                        <h3>\n",
       "                            1&emsp;On the stability of sequences inserted into viral genomes \n",
       "                        </h3>\n",
       "                        <cite>\n",
       "                            doi\n",
       "                            <span >\n",
       "                                > 10.1093/ve/vez045\n",
       "                            </span>\n",
       "                        </cite>\n",
       "                    </div>\n",
       "                </a>\n",
       "        \n",
       "                <li>\n",
       "                <a href=\" https://doi.org/10.3389/fbioe.2019.00110 \" target=\"_blank\">\n",
       "                    <div>\n",
       "                        <h3>\n",
       "                            2&emsp;Microbial Proteases Applications \n",
       "                        </h3>\n",
       "                        <cite>\n",
       "                            doi\n",
       "                            <span >\n",
       "                                > 10.3389/fbioe.2019.00110\n",
       "                            </span>\n",
       "                        </cite>\n",
       "                    </div>\n",
       "                </a>\n",
       "        \n",
       "                <li>\n",
       "                <a href=\" https://doi.org/10.1101/358564 \" target=\"_blank\">\n",
       "                    <div>\n",
       "                        <h3>\n",
       "                            3&emsp;Changes in temperature alter susceptibility to a virus following a host shift \n",
       "                        </h3>\n",
       "                        <cite>\n",
       "                            doi\n",
       "                            <span >\n",
       "                                > 10.1101/358564\n",
       "                            </span>\n",
       "                        </cite>\n",
       "                    </div>\n",
       "                </a>\n",
       "        \n",
       "                <li>\n",
       "                <a href=\" https://doi.org/10.1016/j.febslet.2012.02.022 \" target=\"_blank\">\n",
       "                    <div>\n",
       "                        <h3>\n",
       "                            4&emsp;Synthetic constructs in/for the environment: Managing the interplay between natural and engineered Biology \n",
       "                        </h3>\n",
       "                        <cite>\n",
       "                            doi\n",
       "                            <span >\n",
       "                                > 10.1016/j.febslet.2012.02.022\n",
       "                            </span>\n",
       "                        </cite>\n",
       "                    </div>\n",
       "                </a>\n",
       "        \n",
       "                <li>\n",
       "                <a href=\" https://doi.org/10.1016/b978-0-12-804177-2.00004-x \" target=\"_blank\">\n",
       "                    <div>\n",
       "                        <h3>\n",
       "                            5&emsp;Chapter 4 Structure and Genotypes of Noroviruses \n",
       "                        </h3>\n",
       "                        <cite>\n",
       "                            doi\n",
       "                            <span >\n",
       "                                > 10.1016/b978-0-12-804177-2.00004-x\n",
       "                            </span>\n",
       "                        </cite>\n",
       "                    </div>\n",
       "                </a>\n",
       "        </body></html>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "query('environmental stability of coronavirus', top_matches=5)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "> # What do we know about diagnostics and surveillance?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "    <html>\n",
       "    <style>\n",
       "        body {\n",
       "            margin: 0;\n",
       "            padding: 0;\n",
       "            background: #ffffff;\n",
       "            font-family: sans-serif;\n",
       "        }\n",
       "\n",
       "        ul {\n",
       "            position: relative;\n",
       "            width: 100%;\n",
       "            margin: 100px auto 0;\n",
       "            padding: 20px;\n",
       "            box-sizing: border-box;\n",
       "            background: rgba(0, 0, 0, 0);\n",
       "            box-shadow: inset 0 0 15px rgba(0, 0, 0, .2);\n",
       "            border-radius: 5px;\n",
       "            overflow: hidden;\n",
       "        }\n",
       "\n",
       "        ul li {\n",
       "            display: flex;\n",
       "            background: rgba(255, 0, 0, 0.25);\n",
       "            padding: 10px 20px;\n",
       "            color: #fff;\n",
       "            margin: 5px 0;\n",
       "            transition: .6s;\n",
       "            border-radius: 5px;\n",
       "        }\n",
       "\n",
       "        ul li:hover {\n",
       "            transform: scale(1.02);\n",
       "            background: rgba(255, 0, 0, 0.5);\n",
       "        }\n",
       "    </style>\n",
       "        \n",
       "        <body>\n",
       "        <ul>\n",
       "    \n",
       "                <li>\n",
       "                <a href=\" https://doi.org/10.1007/978-3-030-03071-1_13 \" target=\"_blank\">\n",
       "                    <div>\n",
       "                        <h3>\n",
       "                            1&emsp;Advances in Molecular Diagnostic Approaches for Biothreat Agents \n",
       "                        </h3>\n",
       "                        <cite>\n",
       "                            doi\n",
       "                            <span >\n",
       "                                > 10.1007/978-3-030-03071-1_13\n",
       "                            </span>\n",
       "                        </cite>\n",
       "                    </div>\n",
       "                </a>\n",
       "        \n",
       "                <li>\n",
       "                <a href=\" https://doi.org/10.1016/j.cll.2009.07.005 \" target=\"_blank\">\n",
       "                    <div>\n",
       "                        <h3>\n",
       "                            2&emsp;Emerging Molecular Assays for Detection and Characterization of Respiratory Viruses \n",
       "                        </h3>\n",
       "                        <cite>\n",
       "                            doi\n",
       "                            <span >\n",
       "                                > 10.1016/j.cll.2009.07.005\n",
       "                            </span>\n",
       "                        </cite>\n",
       "                    </div>\n",
       "                </a>\n",
       "        \n",
       "                <li>\n",
       "                <a href=\" https://doi.org/10.1007/978-1-4614-3970-7_11 \" target=\"_blank\">\n",
       "                    <div>\n",
       "                        <h3>\n",
       "                            3&emsp;Nucleic Acid Extraction Techniques \n",
       "                        </h3>\n",
       "                        <cite>\n",
       "                            doi\n",
       "                            <span >\n",
       "                                > 10.1007/978-1-4614-3970-7_11\n",
       "                            </span>\n",
       "                        </cite>\n",
       "                    </div>\n",
       "                </a>\n",
       "        \n",
       "                <li>\n",
       "                <a href=\" https://doi.org/10.1053/j.jepm.2010.05.006 \" target=\"_blank\">\n",
       "                    <div>\n",
       "                        <h3>\n",
       "                            4&emsp;Future Veterinary Diagnostics \n",
       "                        </h3>\n",
       "                        <cite>\n",
       "                            doi\n",
       "                            <span >\n",
       "                                > 10.1053/j.jepm.2010.05.006\n",
       "                            </span>\n",
       "                        </cite>\n",
       "                    </div>\n",
       "                </a>\n",
       "        \n",
       "                <li>\n",
       "                <a href=\" https://doi.org/10.1016/j.cvsm.2010.07.012 \" target=\"_blank\">\n",
       "                    <div>\n",
       "                        <h3>\n",
       "                            5&emsp;Molecular Diagnostic Assays for Infectious Diseases in Cats \n",
       "                        </h3>\n",
       "                        <cite>\n",
       "                            doi\n",
       "                            <span >\n",
       "                                > 10.1016/j.cvsm.2010.07.012\n",
       "                            </span>\n",
       "                        </cite>\n",
       "                    </div>\n",
       "                </a>\n",
       "        \n",
       "                <li>\n",
       "                <a href=\" https://doi.org/10.3390/diagnostics7030039 \" target=\"_blank\">\n",
       "                    <div>\n",
       "                        <h3>\n",
       "                            6&emsp;Advancing Point-of-Care (PoC) Testing Using Human Saliva as Liquid Biopsy \n",
       "                        </h3>\n",
       "                        <cite>\n",
       "                            doi\n",
       "                            <span >\n",
       "                                > 10.3390/diagnostics7030039\n",
       "                            </span>\n",
       "                        </cite>\n",
       "                    </div>\n",
       "                </a>\n",
       "        \n",
       "                <li>\n",
       "                <a href=\" https://doi.org/10.1007/bf01839180 \" target=\"_blank\">\n",
       "                    <div>\n",
       "                        <h3>\n",
       "                            7&emsp;Application of the polymerase chain reaction (PCR) in veterinary diagnostic virology \n",
       "                        </h3>\n",
       "                        <cite>\n",
       "                            doi\n",
       "                            <span >\n",
       "                                > 10.1007/bf01839180\n",
       "                            </span>\n",
       "                        </cite>\n",
       "                    </div>\n",
       "                </a>\n",
       "        \n",
       "                <li>\n",
       "                <a href=\" https://doi.org/10.1016/b978-0-12-816352-8.00013-8 \" target=\"_blank\">\n",
       "                    <div>\n",
       "                        <h3>\n",
       "                            8&emsp;Chapter 13 Biotechnological innovations in farm and pet animal disease diagnosis \n",
       "                        </h3>\n",
       "                        <cite>\n",
       "                            doi\n",
       "                            <span >\n",
       "                                > 10.1016/b978-0-12-816352-8.00013-8\n",
       "                            </span>\n",
       "                        </cite>\n",
       "                    </div>\n",
       "                </a>\n",
       "        \n",
       "                <li>\n",
       "                <a href=\" https://doi.org/10.1097/pat.0000000000000240 \" target=\"_blank\">\n",
       "                    <div>\n",
       "                        <h3>\n",
       "                            9&emsp;Molecular diagnosis of respiratory viruses \n",
       "                        </h3>\n",
       "                        <cite>\n",
       "                            doi\n",
       "                            <span >\n",
       "                                > 10.1097/pat.0000000000000240\n",
       "                            </span>\n",
       "                        </cite>\n",
       "                    </div>\n",
       "                </a>\n",
       "        \n",
       "                <li>\n",
       "                <a href=\" https://doi.org/10.1007/s10156-009-0669-9 \" target=\"_blank\">\n",
       "                    <div>\n",
       "                        <h3>\n",
       "                            10&emsp;Loop-mediated isothermal amplification (LAMP): a rapid, accurate, and cost-effective diagnostic method for infectious diseases \n",
       "                        </h3>\n",
       "                        <cite>\n",
       "                            doi\n",
       "                            <span >\n",
       "                                > 10.1007/s10156-009-0669-9\n",
       "                            </span>\n",
       "                        </cite>\n",
       "                    </div>\n",
       "                </a>\n",
       "        </body></html>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "query('diagnostics')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Thank you for your time, your reviews and suggenstions are highly appreciated !"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "nlp",
   "language": "python",
   "name": "nlp"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.10"
  },
  "widgets": {
   "application/vnd.jupyter.widget-state+json": {
    "state": {
     "03a9c4865c0f4f6daeb421ac227d5383": {
      "model_module": "@jupyter-widgets/controls",
      "model_module_version": "1.5.0",
      "model_name": "ProgressStyleModel",
      "state": {
       "_model_module": "@jupyter-widgets/controls",
       "_model_module_version": "1.5.0",
       "_model_name": "ProgressStyleModel",
       "_view_count": null,
       "_view_module": "@jupyter-widgets/base",
       "_view_module_version": "1.2.0",
       "_view_name": "StyleView",
       "bar_color": null,
       "description_width": "initial"
      }
     },
     "32af0d3a298844c8acbe1288db35db7b": {
      "model_module": "@jupyter-widgets/controls",
      "model_module_version": "1.5.0",
      "model_name": "DescriptionStyleModel",
      "state": {
       "_model_module": "@jupyter-widgets/controls",
       "_model_module_version": "1.5.0",
       "_model_name": "DescriptionStyleModel",
       "_view_count": null,
       "_view_module": "@jupyter-widgets/base",
       "_view_module_version": "1.2.0",
       "_view_name": "StyleView",
       "description_width": ""
      }
     },
     "336deea24d6a4e83a4680fdcfb069643": {
      "model_module": "@jupyter-widgets/base",
      "model_module_version": "1.2.0",
      "model_name": "LayoutModel",
      "state": {
       "_model_module": "@jupyter-widgets/base",
       "_model_module_version": "1.2.0",
       "_model_name": "LayoutModel",
       "_view_count": null,
       "_view_module": "@jupyter-widgets/base",
       "_view_module_version": "1.2.0",
       "_view_name": "LayoutView",
       "align_content": null,
       "align_items": null,
       "align_self": null,
       "border": null,
       "bottom": null,
       "display": null,
       "flex": null,
       "flex_flow": null,
       "grid_area": null,
       "grid_auto_columns": null,
       "grid_auto_flow": null,
       "grid_auto_rows": null,
       "grid_column": null,
       "grid_gap": null,
       "grid_row": null,
       "grid_template_areas": null,
       "grid_template_columns": null,
       "grid_template_rows": null,
       "height": null,
       "justify_content": null,
       "justify_items": null,
       "left": null,
       "margin": null,
       "max_height": null,
       "max_width": null,
       "min_height": null,
       "min_width": null,
       "object_fit": null,
       "object_position": null,
       "order": null,
       "overflow": null,
       "overflow_x": null,
       "overflow_y": null,
       "padding": null,
       "right": null,
       "top": null,
       "visibility": null,
       "width": null
      }
     },
     "39b0e580ee114f0f86b33e3918d2a917": {
      "model_module": "@jupyter-widgets/controls",
      "model_module_version": "1.5.0",
      "model_name": "FloatProgressModel",
      "state": {
       "_dom_classes": [],
       "_model_module": "@jupyter-widgets/controls",
       "_model_module_version": "1.5.0",
       "_model_name": "FloatProgressModel",
       "_view_count": null,
       "_view_module": "@jupyter-widgets/controls",
       "_view_module_version": "1.5.0",
       "_view_name": "ProgressView",
       "bar_style": "success",
       "description": "100%",
       "description_tooltip": null,
       "layout": "IPY_MODEL_f8c356af8f4e46d8ae6650283c0fc51c",
       "max": 38071,
       "min": 0,
       "orientation": "horizontal",
       "style": "IPY_MODEL_a5989e87bcd741ea9e2fa2d6737baf95",
       "value": 38071
      }
     },
     "39d346bd287d4706829b16feec4bc090": {
      "model_module": "@jupyter-widgets/controls",
      "model_module_version": "1.5.0",
      "model_name": "HBoxModel",
      "state": {
       "_dom_classes": [],
       "_model_module": "@jupyter-widgets/controls",
       "_model_module_version": "1.5.0",
       "_model_name": "HBoxModel",
       "_view_count": null,
       "_view_module": "@jupyter-widgets/controls",
       "_view_module_version": "1.5.0",
       "_view_name": "HBoxView",
       "box_style": "",
       "children": [
        "IPY_MODEL_39b0e580ee114f0f86b33e3918d2a917",
        "IPY_MODEL_e5868906c0974a259bf5eb90a0de7f78"
       ],
       "layout": "IPY_MODEL_417eec2ef7e04794a042afbc17002a46"
      }
     },
     "3d5a0c5b0bea4d26ae6bd3e0481925b9": {
      "model_module": "@jupyter-widgets/base",
      "model_module_version": "1.2.0",
      "model_name": "LayoutModel",
      "state": {
       "_model_module": "@jupyter-widgets/base",
       "_model_module_version": "1.2.0",
       "_model_name": "LayoutModel",
       "_view_count": null,
       "_view_module": "@jupyter-widgets/base",
       "_view_module_version": "1.2.0",
       "_view_name": "LayoutView",
       "align_content": null,
       "align_items": null,
       "align_self": null,
       "border": null,
       "bottom": null,
       "display": null,
       "flex": null,
       "flex_flow": null,
       "grid_area": null,
       "grid_auto_columns": null,
       "grid_auto_flow": null,
       "grid_auto_rows": null,
       "grid_column": null,
       "grid_gap": null,
       "grid_row": null,
       "grid_template_areas": null,
       "grid_template_columns": null,
       "grid_template_rows": null,
       "height": null,
       "justify_content": null,
       "justify_items": null,
       "left": null,
       "margin": null,
       "max_height": null,
       "max_width": null,
       "min_height": null,
       "min_width": null,
       "object_fit": null,
       "object_position": null,
       "order": null,
       "overflow": null,
       "overflow_x": null,
       "overflow_y": null,
       "padding": null,
       "right": null,
       "top": null,
       "visibility": null,
       "width": null
      }
     },
     "3ed529705dff4da9835484356a349c10": {
      "model_module": "@jupyter-widgets/base",
      "model_module_version": "1.2.0",
      "model_name": "LayoutModel",
      "state": {
       "_model_module": "@jupyter-widgets/base",
       "_model_module_version": "1.2.0",
       "_model_name": "LayoutModel",
       "_view_count": null,
       "_view_module": "@jupyter-widgets/base",
       "_view_module_version": "1.2.0",
       "_view_name": "LayoutView",
       "align_content": null,
       "align_items": null,
       "align_self": null,
       "border": null,
       "bottom": null,
       "display": null,
       "flex": null,
       "flex_flow": null,
       "grid_area": null,
       "grid_auto_columns": null,
       "grid_auto_flow": null,
       "grid_auto_rows": null,
       "grid_column": null,
       "grid_gap": null,
       "grid_row": null,
       "grid_template_areas": null,
       "grid_template_columns": null,
       "grid_template_rows": null,
       "height": null,
       "justify_content": null,
       "justify_items": null,
       "left": null,
       "margin": null,
       "max_height": null,
       "max_width": null,
       "min_height": null,
       "min_width": null,
       "object_fit": null,
       "object_position": null,
       "order": null,
       "overflow": null,
       "overflow_x": null,
       "overflow_y": null,
       "padding": null,
       "right": null,
       "top": null,
       "visibility": null,
       "width": null
      }
     },
     "417eec2ef7e04794a042afbc17002a46": {
      "model_module": "@jupyter-widgets/base",
      "model_module_version": "1.2.0",
      "model_name": "LayoutModel",
      "state": {
       "_model_module": "@jupyter-widgets/base",
       "_model_module_version": "1.2.0",
       "_model_name": "LayoutModel",
       "_view_count": null,
       "_view_module": "@jupyter-widgets/base",
       "_view_module_version": "1.2.0",
       "_view_name": "LayoutView",
       "align_content": null,
       "align_items": null,
       "align_self": null,
       "border": null,
       "bottom": null,
       "display": null,
       "flex": null,
       "flex_flow": null,
       "grid_area": null,
       "grid_auto_columns": null,
       "grid_auto_flow": null,
       "grid_auto_rows": null,
       "grid_column": null,
       "grid_gap": null,
       "grid_row": null,
       "grid_template_areas": null,
       "grid_template_columns": null,
       "grid_template_rows": null,
       "height": null,
       "justify_content": null,
       "justify_items": null,
       "left": null,
       "margin": null,
       "max_height": null,
       "max_width": null,
       "min_height": null,
       "min_width": null,
       "object_fit": null,
       "object_position": null,
       "order": null,
       "overflow": null,
       "overflow_x": null,
       "overflow_y": null,
       "padding": null,
       "right": null,
       "top": null,
       "visibility": null,
       "width": null
      }
     },
     "48783991ff464088a61863cab41507ea": {
      "model_module": "@jupyter-widgets/controls",
      "model_module_version": "1.5.0",
      "model_name": "DescriptionStyleModel",
      "state": {
       "_model_module": "@jupyter-widgets/controls",
       "_model_module_version": "1.5.0",
       "_model_name": "DescriptionStyleModel",
       "_view_count": null,
       "_view_module": "@jupyter-widgets/base",
       "_view_module_version": "1.2.0",
       "_view_name": "StyleView",
       "description_width": ""
      }
     },
     "5007ec5d80f2476e9a94386cc61f3c24": {
      "model_module": "@jupyter-widgets/controls",
      "model_module_version": "1.5.0",
      "model_name": "HTMLModel",
      "state": {
       "_dom_classes": [],
       "_model_module": "@jupyter-widgets/controls",
       "_model_module_version": "1.5.0",
       "_model_name": "HTMLModel",
       "_view_count": null,
       "_view_module": "@jupyter-widgets/controls",
       "_view_module_version": "1.5.0",
       "_view_name": "HTMLView",
       "description": "",
       "description_tooltip": null,
       "layout": "IPY_MODEL_fce38a60c4fd4e40a79364bceae7e8e9",
       "placeholder": "​",
       "style": "IPY_MODEL_48783991ff464088a61863cab41507ea",
       "value": " 38584/38584 [09:51&lt;00:00, 65.23it/s]"
      }
     },
     "5491ae7c37a64f829848c9718bac415e": {
      "model_module": "@jupyter-widgets/base",
      "model_module_version": "1.2.0",
      "model_name": "LayoutModel",
      "state": {
       "_model_module": "@jupyter-widgets/base",
       "_model_module_version": "1.2.0",
       "_model_name": "LayoutModel",
       "_view_count": null,
       "_view_module": "@jupyter-widgets/base",
       "_view_module_version": "1.2.0",
       "_view_name": "LayoutView",
       "align_content": null,
       "align_items": null,
       "align_self": null,
       "border": null,
       "bottom": null,
       "display": null,
       "flex": null,
       "flex_flow": null,
       "grid_area": null,
       "grid_auto_columns": null,
       "grid_auto_flow": null,
       "grid_auto_rows": null,
       "grid_column": null,
       "grid_gap": null,
       "grid_row": null,
       "grid_template_areas": null,
       "grid_template_columns": null,
       "grid_template_rows": null,
       "height": null,
       "justify_content": null,
       "justify_items": null,
       "left": null,
       "margin": null,
       "max_height": null,
       "max_width": null,
       "min_height": null,
       "min_width": null,
       "object_fit": null,
       "object_position": null,
       "order": null,
       "overflow": null,
       "overflow_x": null,
       "overflow_y": null,
       "padding": null,
       "right": null,
       "top": null,
       "visibility": null,
       "width": null
      }
     },
     "5514bdb5f1724820a9882927457e6c28": {
      "model_module": "@jupyter-widgets/controls",
      "model_module_version": "1.5.0",
      "model_name": "ProgressStyleModel",
      "state": {
       "_model_module": "@jupyter-widgets/controls",
       "_model_module_version": "1.5.0",
       "_model_name": "ProgressStyleModel",
       "_view_count": null,
       "_view_module": "@jupyter-widgets/base",
       "_view_module_version": "1.2.0",
       "_view_name": "StyleView",
       "bar_color": null,
       "description_width": "initial"
      }
     },
     "56eb7c5e16294a4e8939c458abea7880": {
      "model_module": "@jupyter-widgets/base",
      "model_module_version": "1.2.0",
      "model_name": "LayoutModel",
      "state": {
       "_model_module": "@jupyter-widgets/base",
       "_model_module_version": "1.2.0",
       "_model_name": "LayoutModel",
       "_view_count": null,
       "_view_module": "@jupyter-widgets/base",
       "_view_module_version": "1.2.0",
       "_view_name": "LayoutView",
       "align_content": null,
       "align_items": null,
       "align_self": null,
       "border": null,
       "bottom": null,
       "display": null,
       "flex": null,
       "flex_flow": null,
       "grid_area": null,
       "grid_auto_columns": null,
       "grid_auto_flow": null,
       "grid_auto_rows": null,
       "grid_column": null,
       "grid_gap": null,
       "grid_row": null,
       "grid_template_areas": null,
       "grid_template_columns": null,
       "grid_template_rows": null,
       "height": null,
       "justify_content": null,
       "justify_items": null,
       "left": null,
       "margin": null,
       "max_height": null,
       "max_width": null,
       "min_height": null,
       "min_width": null,
       "object_fit": null,
       "object_position": null,
       "order": null,
       "overflow": null,
       "overflow_x": null,
       "overflow_y": null,
       "padding": null,
       "right": null,
       "top": null,
       "visibility": null,
       "width": null
      }
     },
     "5b4c3c748381473fb3749bb65a64c4f6": {
      "model_module": "@jupyter-widgets/controls",
      "model_module_version": "1.5.0",
      "model_name": "DescriptionStyleModel",
      "state": {
       "_model_module": "@jupyter-widgets/controls",
       "_model_module_version": "1.5.0",
       "_model_name": "DescriptionStyleModel",
       "_view_count": null,
       "_view_module": "@jupyter-widgets/base",
       "_view_module_version": "1.2.0",
       "_view_name": "StyleView",
       "description_width": ""
      }
     },
     "6ef0e0ac4d574941b51a7d893e133c7f": {
      "model_module": "@jupyter-widgets/controls",
      "model_module_version": "1.5.0",
      "model_name": "HBoxModel",
      "state": {
       "_dom_classes": [],
       "_model_module": "@jupyter-widgets/controls",
       "_model_module_version": "1.5.0",
       "_model_name": "HBoxModel",
       "_view_count": null,
       "_view_module": "@jupyter-widgets/controls",
       "_view_module_version": "1.5.0",
       "_view_name": "HBoxView",
       "box_style": "",
       "children": [
        "IPY_MODEL_8194cb0343f04a0290c06a59bb175856",
        "IPY_MODEL_ecf1f9d315d14657a7694d2e830b2d1c"
       ],
       "layout": "IPY_MODEL_5491ae7c37a64f829848c9718bac415e"
      }
     },
     "726a1c446a1742699dfd08cf9ca1864b": {
      "model_module": "@jupyter-widgets/base",
      "model_module_version": "1.2.0",
      "model_name": "LayoutModel",
      "state": {
       "_model_module": "@jupyter-widgets/base",
       "_model_module_version": "1.2.0",
       "_model_name": "LayoutModel",
       "_view_count": null,
       "_view_module": "@jupyter-widgets/base",
       "_view_module_version": "1.2.0",
       "_view_name": "LayoutView",
       "align_content": null,
       "align_items": null,
       "align_self": null,
       "border": null,
       "bottom": null,
       "display": null,
       "flex": null,
       "flex_flow": null,
       "grid_area": null,
       "grid_auto_columns": null,
       "grid_auto_flow": null,
       "grid_auto_rows": null,
       "grid_column": null,
       "grid_gap": null,
       "grid_row": null,
       "grid_template_areas": null,
       "grid_template_columns": null,
       "grid_template_rows": null,
       "height": null,
       "justify_content": null,
       "justify_items": null,
       "left": null,
       "margin": null,
       "max_height": null,
       "max_width": null,
       "min_height": null,
       "min_width": null,
       "object_fit": null,
       "object_position": null,
       "order": null,
       "overflow": null,
       "overflow_x": null,
       "overflow_y": null,
       "padding": null,
       "right": null,
       "top": null,
       "visibility": null,
       "width": null
      }
     },
     "8194cb0343f04a0290c06a59bb175856": {
      "model_module": "@jupyter-widgets/controls",
      "model_module_version": "1.5.0",
      "model_name": "FloatProgressModel",
      "state": {
       "_dom_classes": [],
       "_model_module": "@jupyter-widgets/controls",
       "_model_module_version": "1.5.0",
       "_model_name": "FloatProgressModel",
       "_view_count": null,
       "_view_module": "@jupyter-widgets/controls",
       "_view_module_version": "1.5.0",
       "_view_name": "ProgressView",
       "bar_style": "",
       "description": " 46%",
       "description_tooltip": null,
       "layout": "IPY_MODEL_336deea24d6a4e83a4680fdcfb069643",
       "max": 38071,
       "min": 0,
       "orientation": "horizontal",
       "style": "IPY_MODEL_03a9c4865c0f4f6daeb421ac227d5383",
       "value": 17620
      }
     },
     "854fef8104aa4bfa9ba033f34545ac30": {
      "model_module": "@jupyter-widgets/controls",
      "model_module_version": "1.5.0",
      "model_name": "HBoxModel",
      "state": {
       "_dom_classes": [],
       "_model_module": "@jupyter-widgets/controls",
       "_model_module_version": "1.5.0",
       "_model_name": "HBoxModel",
       "_view_count": null,
       "_view_module": "@jupyter-widgets/controls",
       "_view_module_version": "1.5.0",
       "_view_name": "HBoxView",
       "box_style": "",
       "children": [
        "IPY_MODEL_a3f1b4c7d8294b2cb1f433ef88cbb20f",
        "IPY_MODEL_5007ec5d80f2476e9a94386cc61f3c24"
       ],
       "layout": "IPY_MODEL_3d5a0c5b0bea4d26ae6bd3e0481925b9"
      }
     },
     "a3f1b4c7d8294b2cb1f433ef88cbb20f": {
      "model_module": "@jupyter-widgets/controls",
      "model_module_version": "1.5.0",
      "model_name": "FloatProgressModel",
      "state": {
       "_dom_classes": [],
       "_model_module": "@jupyter-widgets/controls",
       "_model_module_version": "1.5.0",
       "_model_name": "FloatProgressModel",
       "_view_count": null,
       "_view_module": "@jupyter-widgets/controls",
       "_view_module_version": "1.5.0",
       "_view_name": "ProgressView",
       "bar_style": "",
       "description": " 70%",
       "description_tooltip": null,
       "layout": "IPY_MODEL_56eb7c5e16294a4e8939c458abea7880",
       "max": 38584,
       "min": 0,
       "orientation": "horizontal",
       "style": "IPY_MODEL_5514bdb5f1724820a9882927457e6c28",
       "value": 27157
      }
     },
     "a5989e87bcd741ea9e2fa2d6737baf95": {
      "model_module": "@jupyter-widgets/controls",
      "model_module_version": "1.5.0",
      "model_name": "ProgressStyleModel",
      "state": {
       "_model_module": "@jupyter-widgets/controls",
       "_model_module_version": "1.5.0",
       "_model_name": "ProgressStyleModel",
       "_view_count": null,
       "_view_module": "@jupyter-widgets/base",
       "_view_module_version": "1.2.0",
       "_view_name": "StyleView",
       "bar_color": null,
       "description_width": "initial"
      }
     },
     "e5868906c0974a259bf5eb90a0de7f78": {
      "model_module": "@jupyter-widgets/controls",
      "model_module_version": "1.5.0",
      "model_name": "HTMLModel",
      "state": {
       "_dom_classes": [],
       "_model_module": "@jupyter-widgets/controls",
       "_model_module_version": "1.5.0",
       "_model_name": "HTMLModel",
       "_view_count": null,
       "_view_module": "@jupyter-widgets/controls",
       "_view_module_version": "1.5.0",
       "_view_name": "HTMLView",
       "description": "",
       "description_tooltip": null,
       "layout": "IPY_MODEL_726a1c446a1742699dfd08cf9ca1864b",
       "placeholder": "​",
       "style": "IPY_MODEL_32af0d3a298844c8acbe1288db35db7b",
       "value": " 38071/38071 [15:17&lt;00:00, 41.51it/s]"
      }
     },
     "ecf1f9d315d14657a7694d2e830b2d1c": {
      "model_module": "@jupyter-widgets/controls",
      "model_module_version": "1.5.0",
      "model_name": "HTMLModel",
      "state": {
       "_dom_classes": [],
       "_model_module": "@jupyter-widgets/controls",
       "_model_module_version": "1.5.0",
       "_model_name": "HTMLModel",
       "_view_count": null,
       "_view_module": "@jupyter-widgets/controls",
       "_view_module_version": "1.5.0",
       "_view_name": "HTMLView",
       "description": "",
       "description_tooltip": null,
       "layout": "IPY_MODEL_3ed529705dff4da9835484356a349c10",
       "placeholder": "​",
       "style": "IPY_MODEL_5b4c3c748381473fb3749bb65a64c4f6",
       "value": " 38071/38071 [48:32&lt;00:00, 13.07it/s]"
      }
     },
     "f8c356af8f4e46d8ae6650283c0fc51c": {
      "model_module": "@jupyter-widgets/base",
      "model_module_version": "1.2.0",
      "model_name": "LayoutModel",
      "state": {
       "_model_module": "@jupyter-widgets/base",
       "_model_module_version": "1.2.0",
       "_model_name": "LayoutModel",
       "_view_count": null,
       "_view_module": "@jupyter-widgets/base",
       "_view_module_version": "1.2.0",
       "_view_name": "LayoutView",
       "align_content": null,
       "align_items": null,
       "align_self": null,
       "border": null,
       "bottom": null,
       "display": null,
       "flex": null,
       "flex_flow": null,
       "grid_area": null,
       "grid_auto_columns": null,
       "grid_auto_flow": null,
       "grid_auto_rows": null,
       "grid_column": null,
       "grid_gap": null,
       "grid_row": null,
       "grid_template_areas": null,
       "grid_template_columns": null,
       "grid_template_rows": null,
       "height": null,
       "justify_content": null,
       "justify_items": null,
       "left": null,
       "margin": null,
       "max_height": null,
       "max_width": null,
       "min_height": null,
       "min_width": null,
       "object_fit": null,
       "object_position": null,
       "order": null,
       "overflow": null,
       "overflow_x": null,
       "overflow_y": null,
       "padding": null,
       "right": null,
       "top": null,
       "visibility": null,
       "width": null
      }
     },
     "fce38a60c4fd4e40a79364bceae7e8e9": {
      "model_module": "@jupyter-widgets/base",
      "model_module_version": "1.2.0",
      "model_name": "LayoutModel",
      "state": {
       "_model_module": "@jupyter-widgets/base",
       "_model_module_version": "1.2.0",
       "_model_name": "LayoutModel",
       "_view_count": null,
       "_view_module": "@jupyter-widgets/base",
       "_view_module_version": "1.2.0",
       "_view_name": "LayoutView",
       "align_content": null,
       "align_items": null,
       "align_self": null,
       "border": null,
       "bottom": null,
       "display": null,
       "flex": null,
       "flex_flow": null,
       "grid_area": null,
       "grid_auto_columns": null,
       "grid_auto_flow": null,
       "grid_auto_rows": null,
       "grid_column": null,
       "grid_gap": null,
       "grid_row": null,
       "grid_template_areas": null,
       "grid_template_columns": null,
       "grid_template_rows": null,
       "height": null,
       "justify_content": null,
       "justify_items": null,
       "left": null,
       "margin": null,
       "max_height": null,
       "max_width": null,
       "min_height": null,
       "min_width": null,
       "object_fit": null,
       "object_position": null,
       "order": null,
       "overflow": null,
       "overflow_x": null,
       "overflow_y": null,
       "padding": null,
       "right": null,
       "top": null,
       "visibility": null,
       "width": null
      }
     }
    },
    "version_major": 2,
    "version_minor": 0
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
